// @kepler.gl/deckgl-layers/dist/deckgl-extensions/filter-arrow-layer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _core = require("@deck.gl/core");

var _constants = _interopRequireDefault(require("@luma.gl/constants"));

var _filterShaderModule = _interopRequireDefault(require("./filter-shader-module"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var VALUE_FILTERED = 1;
var defaultProps = {
  getFiltered: {
    type: 'accessor',
    value: VALUE_FILTERED
  }
};

/**
 * FilterArrowExtension - a deck.gl extension to filter arrow layer
 *
 * A simple extension to filter arrow layer based on the result of CPU filteredIndex,
 * so we can avoid filtering on the raw Arrow table and recreating geometry attributes.
 * Specifically, an attribute `filtered` is added to the layer to indicate whether the feature has been Filtered
 * the shader module is modified to discard the feature if filtered value is 0
 * the accessor getFiltered is used to get the value of `filtered` based on the value `filteredIndex` in Arrowlayer
 */
var FilterArrowExtension = /*#__PURE__*/function (_LayerExtension) {
  (0, _inherits2["default"])(FilterArrowExtension, _LayerExtension);

  var _super = _createSuper(FilterArrowExtension);

  function FilterArrowExtension() {
    (0, _classCallCheck2["default"])(this, FilterArrowExtension);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(FilterArrowExtension, [{
    key: "getShaders",
    value: function getShaders(extension) {
      return {
        modules: [_filterShaderModule["default"]],
        defines: {}
      };
    }
  }, {
    key: "initializeState",
    value: function initializeState(context, extension) {
      var attributeManager = this.getAttributeManager();

      if (attributeManager) {
        attributeManager.add({
          filtered: {
            size: 1,
            type: _constants["default"].FLOAT,
            accessor: 'getFiltered',
            shaderAttributes: {
              filtered: {
                divisor: 0
              },
              instanceFiltered: {
                divisor: 1
              }
            }
          }
        });
      }
    }
  }]);
  return FilterArrowExtension;
}(_core.LayerExtension);

exports["default"] = FilterArrowExtension;
(0, _defineProperty2["default"])(FilterArrowExtension, "defaultProps", defaultProps);
(0, _defineProperty2["default"])(FilterArrowExtension, "extensionName", 'FilterArrowExtension');
