// @hubble.gl/core/dist/es6/encoders/video/jpeg-sequence-encoder.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import FrameEncoder from '../frame-encoder';
import TARBuilder from '../tar/tar-builder';
import { pad, canvasToArrayBuffer } from '../utils';

class JPEGSequenceEncoder extends FrameEncoder {
  constructor(settings) {
    super(settings);

    _defineProperty(this, "tarBuilder", void 0);

    this.mimeType = TARBuilder.properties.mimeType;
    this.extension = ".".concat(TARBuilder.properties.extensions[0]);
    this.quality = 0.8;

    if (settings.jpeg && settings.jpeg.quality) {
      this.quality = settings.jpeg.quality;
    }

    this.tarBuilder = null;
  }

  start() {
    this.tarBuilder = new TARBuilder({});
  }

  async add(canvas) {
    const mimeType = 'image/jpeg';
    const extension = '.jpg';
    const buffer = await canvasToArrayBuffer(canvas, mimeType);
    const filename = pad(this.tarBuilder.count) + extension;
    this.tarBuilder.addFile(buffer, filename);
  }

  async save() {
    return this.tarBuilder.build();
  }

}

export default JPEGSequenceEncoder;
