// @hubble.gl/core/dist/es6/encoders/video/webm-encoder.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import WebMWriter from 'webm-writer';
import FrameEncoder from '../frame-encoder';

class WebMEncoder extends FrameEncoder {
  constructor(settings) {
    super(settings);

    _defineProperty(this, "videoWriter", void 0);

    this.quality = 0.8;

    if (settings.webm && settings.webm.quality) {
      this.quality = settings.webm.quality;
    }

    const canvas = document.createElement('canvas');

    if (canvas.toDataURL('image/webp').substr(5, 10) !== 'image/webp') {
      console.error('WebP not supported - try another export format');
    }

    this.extension = '.webm';
    this.mimeType = 'video/webm';
    this.videoWriter = null;
    this.start = this.start.bind(this);
    this.add = this.add.bind(this);
    this.save = this.save.bind(this);
  }

  start() {
    this.videoWriter = new WebMWriter({
      quality: this.quality,
      fileWriter: null,
      fd: null,
      frameRate: this.framerate
    });
  }

  async add(canvas) {
    this.videoWriter.addFrame(canvas);
    await Promise.resolve();
  }

  async save() {
    return this.videoWriter.complete();
  }

}

export default WebMEncoder;
