// kepler.gl/src/effects/dist/effect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.Effect = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _utils = require("@kepler.gl/utils");

var _constants = require("@kepler.gl/constants");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var Effect = /*#__PURE__*/function () {
  // effect specific parameters for a deck.gl effect (uniforms)
  // runtime props
  function Effect() {
    var _POSTPROCESSING_EFFEC;

    var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck2["default"])(this, Effect);
    (0, _defineProperty2["default"])(this, "id", void 0);
    (0, _defineProperty2["default"])(this, "type", void 0);
    (0, _defineProperty2["default"])(this, "isEnabled", void 0);
    (0, _defineProperty2["default"])(this, "parameters", void 0);
    (0, _defineProperty2["default"])(this, "isConfigActive", void 0);
    (0, _defineProperty2["default"])(this, "isJsonEditorActive", void 0);
    (0, _defineProperty2["default"])(this, "deckEffect", void 0);
    (0, _defineProperty2["default"])(this, "_uiConfig", void 0);
    this.id = props.id || "e_".concat((0, _utils.generateHashId)(6));

    var _props = this.getDefaultProps(props);

    this.type = _props.type;
    this.isEnabled = _props.isEnabled;
    this.isConfigActive = _props.isConfigActive;
    this.isJsonEditorActive = _props.isJsonEditorActive;
    this._uiConfig = _constants.LIGHT_AND_SHADOW_EFFECT.type === this.type ? _constants.LIGHT_AND_SHADOW_EFFECT.parameters : ((_POSTPROCESSING_EFFEC = _constants.POSTPROCESSING_EFFECTS[this.type]) === null || _POSTPROCESSING_EFFEC === void 0 ? void 0 : _POSTPROCESSING_EFFEC.parameters) || [];
    this.parameters = (0, _utils.validateEffectParameters)(_props.parameters, this._uiConfig);
    this.deckEffect = null;

    this._initializeEffect();
  }

  (0, _createClass2["default"])(Effect, [{
    key: "_initializeEffect",
    value: function _initializeEffect() {// implemented in subclasses
    }
  }, {
    key: "getDefaultProps",
    value: function getDefaultProps() {
      var _props$isEnabled, _props$isConfigActive, _props$isJsonEditorAc;

      var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      return {
        id: props.id || "e_".concat((0, _utils.generateHashId)(6)),
        type: props.type || _constants.DEFAULT_POST_PROCESSING_EFFECT_TYPE,
        isEnabled: (_props$isEnabled = props.isEnabled) !== null && _props$isEnabled !== void 0 ? _props$isEnabled : true,
        isConfigActive: (_props$isConfigActive = props.isConfigActive) !== null && _props$isConfigActive !== void 0 ? _props$isConfigActive : true,
        isJsonEditorActive: (_props$isJsonEditorAc = props.isJsonEditorActive) !== null && _props$isJsonEditorAc !== void 0 ? _props$isJsonEditorAc : false,
        parameters: _objectSpread({}, props.parameters)
      };
    }
  }, {
    key: "setProps",
    value: function setProps(props) {
      var _props$id, _props$type, _props$isEnabled2, _props$isConfigActive2, _props$isJsonEditorAc2;

      this.id = (_props$id = props.id) !== null && _props$id !== void 0 ? _props$id : this.id;
      this.type = (_props$type = props.type) !== null && _props$type !== void 0 ? _props$type : this.type;
      this.isEnabled = (_props$isEnabled2 = props.isEnabled) !== null && _props$isEnabled2 !== void 0 ? _props$isEnabled2 : this.isEnabled;
      this.isConfigActive = (_props$isConfigActive2 = props.isConfigActive) !== null && _props$isConfigActive2 !== void 0 ? _props$isConfigActive2 : this.isConfigActive;
      this.isJsonEditorActive = (_props$isJsonEditorAc2 = props.isJsonEditorActive) !== null && _props$isJsonEditorAc2 !== void 0 ? _props$isJsonEditorAc2 : this.isJsonEditorActive;
      this.parameters = _objectSpread(_objectSpread({}, this.parameters), (0, _utils.validateEffectParameters)(props.parameters, this._uiConfig));
    }
  }, {
    key: "isValidToSave",
    value: function isValidToSave() {
      return Boolean(this.type && this.id && this.deckEffect);
    }
    /**
     * Effect specific list of configurable parameters.
     * @returns All parameters are in preffered order.
     */

  }, {
    key: "getParameterDescriptions",
    value: function getParameterDescriptions() {
      return this._uiConfig || [];
    }
  }]);
  return Effect;
}();

exports.Effect = Effect;
var _default = Effect;
exports["default"] = _default;
