// @hubble.gl/core/dist/esm/animations/animation-manager.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { Timeline } from '@luma.gl/engine';
export default class AnimationManager {
  constructor(_ref) {
    let {
      timeline = undefined,
      animations = []
    } = _ref;

    _defineProperty(this, "timeline", void 0);

    _defineProperty(this, "animations", {});

    this.timeline = timeline || new Timeline();

    for (const animation of animations) {
      this.attachAnimation(animation);
    }
  }

  attachAnimation(animation) {
    animation.attachKeyframes(this.timeline);
    this.animations[animation.id] = animation;
  }

  setKeyframes(animationId, params) {
    this.animations[animationId].setKeyframes({
      timeline: this.timeline,
      ...params
    });
  }

  getKeyframes(animationId) {
    return this.animations[animationId].getKeyframes();
  }

  getAnimation(animationId) {
    return this.animations[animationId];
  }

  draw() {
    Object.values(this.animations).forEach(animation => animation.draw());
  }

}
