// @hubble.gl/core/dist/es5/encoders/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FrameEncoder", {
  enumerable: true,
  get: function get() {
    return _frameEncoder.default;
  }
});
Object.defineProperty(exports, "GifEncoder", {
  enumerable: true,
  get: function get() {
    return _gifEncoder.default;
  }
});
Object.defineProperty(exports, "JPEGEncoder", {
  enumerable: true,
  get: function get() {
    return _jpegEncoder.default;
  }
});
Object.defineProperty(exports, "JPEGSequenceEncoder", {
  enumerable: true,
  get: function get() {
    return _jpegSequenceEncoder.default;
  }
});
Object.defineProperty(exports, "PNGEncoder", {
  enumerable: true,
  get: function get() {
    return _pngEncoder.default;
  }
});
Object.defineProperty(exports, "PNGSequenceEncoder", {
  enumerable: true,
  get: function get() {
    return _pngSequenceEncoder.default;
  }
});
Object.defineProperty(exports, "PreviewEncoder", {
  enumerable: true,
  get: function get() {
    return _previewEncoder.default;
  }
});
Object.defineProperty(exports, "StreamEncoder", {
  enumerable: true,
  get: function get() {
    return _streamEncoder.default;
  }
});
Object.defineProperty(exports, "WebMEncoder", {
  enumerable: true,
  get: function get() {
    return _webmEncoder.default;
  }
});

var _pngSequenceEncoder = _interopRequireDefault(require("./video/png-sequence-encoder"));

var _jpegSequenceEncoder = _interopRequireDefault(require("./video/jpeg-sequence-encoder"));

var _jpegEncoder = _interopRequireDefault(require("./photo/jpeg-encoder"));

var _pngEncoder = _interopRequireDefault(require("./photo/png-encoder"));

var _webmEncoder = _interopRequireDefault(require("./video/webm-encoder"));

var _streamEncoder = _interopRequireDefault(require("./video/stream-encoder"));

var _frameEncoder = _interopRequireDefault(require("./frame-encoder"));

var _previewEncoder = _interopRequireDefault(require("./utils/preview-encoder"));

var _gifEncoder = _interopRequireDefault(require("./video/gif-encoder"));
