// @kepler.gl/components/dist/effects/timezone-selector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _momentTimezone = _interopRequireDefault(require("moment-timezone"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _constants = require("@kepler.gl/constants");

var _itemSelector = _interopRequireDefault(require("../common/item-selector/item-selector"));

var _templateObject;

var getValue = function getValue(op) {
  return (op === null || op === void 0 ? void 0 : op.name) || _constants.DEFAULT_TIMEZONE;
};

var StyledItemSelector = (0, _styledComponents["default"])(_itemSelector["default"])(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  .item-selector__dropdown {\n    padding-left: 30px;\n    border-radius: 4px;\n  }\n  .active {\n    border-color: ", ";\n    border-radius: 4px 4px 0px 0px !important;\n  }\n"])), function (props) {
  return props.theme.activeColor;
});

var defaultTimezones = _momentTimezone["default"].tz.names().map(function (name) {
  return {
    name: name
  };
});

function TimezoneSelectorFactory() {
  var TimezoneTypeSelector = function TimezoneTypeSelector(_ref) {
    var selected = _ref.selected,
        _ref$options = _ref.options,
        options = _ref$options === void 0 ? defaultTimezones : _ref$options,
        onSelect = _ref.onSelect;
    var selectedItems = (0, _react.useMemo)(function () {
      return selected ? [{
        name: selected
      }] : [];
    }, [selected]);
    return /*#__PURE__*/_react["default"].createElement(StyledItemSelector, {
      selectedItems: selectedItems,
      options: options,
      multiSelect: false,
      disabled: false,
      placeholder: "effectManager.timezone",
      onChange: onSelect,
      filterOption: "name",
      getOptionValue: getValue,
      displayOption: getValue,
      searchable: true,
      showArrow: true
    });
  };

  return (0, _styledComponents.withTheme)(TimezoneTypeSelector);
}

var _default = TimezoneSelectorFactory;
exports["default"] = _default;
