// kepler.gl/src/utils/dist/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  hexToRgb: true,
  isHexColor: true,
  rgbToHex: true,
  getColorGroupByName: true,
  reverseColorRange: true,
  createLinearGradient: true,
  colorMaybeToRGB: true,
  isRgbColor: true,
  normalizeColor: true,
  errorNotification: true,
  createNotification: true,
  exportImageError: true,
  successNotification: true,
  setStyleSheetBaseHref: true,
  domtoimage: true,
  getFrequency: true,
  getMode: true,
  aggregate: true,
  getTimelineFromAnimationConfig: true,
  getTimelineFromFilter: true,
  SAMPLE_TIMELINE: true,
  TIMELINE_MODES: true,
  datasetColorMaker: true,
  findDefaultColorField: true,
  ACCEPTED_ANALYZER_TYPES: true,
  validateInputData: true,
  getSampleForTypeAnalyze: true,
  getFieldsFromData: true,
  renameDuplicateFields: true,
  analyzerTypeToFieldType: true,
  getFormatLabels: true,
  getFieldFormatLabels: true,
  getFormatValue: true,
  exportMapToHTML: true,
  DEFAULT_IMAGE_NAME: true,
  DEFAULT_HTML_NAME: true,
  DEFAULT_JSON_NAME: true,
  DEFAULT_DATA_NAME: true,
  DEFAULT_EXPORT_JSON_SETTINGS: true,
  isMSEdge: true,
  getScaleFromImageSize: true,
  calculateExportImageSize: true,
  convertToPng: true,
  dataURItoBlob: true,
  downloadFile: true,
  exportImage: true,
  exportToJsonString: true,
  getMapJSON: true,
  exportJson: true,
  exportHtml: true,
  exportMap: true,
  exporters: true,
  setLayerBlending: true,
  flattenMessages: true,
  mergeMessages: true,
  validateToken: true,
  getDefaultLayerGroupVisibility: true,
  editTopMapStyle: true,
  editBottomMapStyle: true,
  isValidStyleUrl: true,
  getStyleDownloadUrl: true,
  getStyleImageIcon: true,
  scaleMapStyleByResolution: true,
  mergeLayerGroupVisibility: true,
  isValidMapInfo: true,
  snapToMarks: true,
  computeDeckEffects: true,
  fixEffectOrder: true,
  reorderEffectOrder: true,
  validateEffectParameters: true,
  transformRequest: true,
  isStyleUsingMapboxTiles: true,
  createDataContainer: true,
  createIndexedDataContainer: true,
  getSampleContainerData: true,
  DataForm: true,
  ArrowDataContainer: true,
  arrowDataTypeToFieldType: true,
  arrowDataTypeToAnalyzerDataType: true,
  getQuantileDomain: true,
  getOrdinalDomain: true,
  getLinearDomain: true,
  getLogDomain: true,
  DataRow: true,
  getCentroid: true,
  idToPolygonGeo: true,
  h3IsValid: true,
  getHexFields: true
};
Object.defineProperty(exports, "hexToRgb", {
  enumerable: true,
  get: function get() {
    return _colorUtils.hexToRgb;
  }
});
Object.defineProperty(exports, "isHexColor", {
  enumerable: true,
  get: function get() {
    return _colorUtils.isHexColor;
  }
});
Object.defineProperty(exports, "rgbToHex", {
  enumerable: true,
  get: function get() {
    return _colorUtils.rgbToHex;
  }
});
Object.defineProperty(exports, "getColorGroupByName", {
  enumerable: true,
  get: function get() {
    return _colorUtils.getColorGroupByName;
  }
});
Object.defineProperty(exports, "reverseColorRange", {
  enumerable: true,
  get: function get() {
    return _colorUtils.reverseColorRange;
  }
});
Object.defineProperty(exports, "createLinearGradient", {
  enumerable: true,
  get: function get() {
    return _colorUtils.createLinearGradient;
  }
});
Object.defineProperty(exports, "colorMaybeToRGB", {
  enumerable: true,
  get: function get() {
    return _colorUtils.colorMaybeToRGB;
  }
});
Object.defineProperty(exports, "isRgbColor", {
  enumerable: true,
  get: function get() {
    return _colorUtils.isRgbColor;
  }
});
Object.defineProperty(exports, "normalizeColor", {
  enumerable: true,
  get: function get() {
    return _colorUtils.normalizeColor;
  }
});
Object.defineProperty(exports, "errorNotification", {
  enumerable: true,
  get: function get() {
    return _notificationsUtils.errorNotification;
  }
});
Object.defineProperty(exports, "createNotification", {
  enumerable: true,
  get: function get() {
    return _notificationsUtils.createNotification;
  }
});
Object.defineProperty(exports, "exportImageError", {
  enumerable: true,
  get: function get() {
    return _notificationsUtils.exportImageError;
  }
});
Object.defineProperty(exports, "successNotification", {
  enumerable: true,
  get: function get() {
    return _notificationsUtils.successNotification;
  }
});
Object.defineProperty(exports, "setStyleSheetBaseHref", {
  enumerable: true,
  get: function get() {
    return _domUtils.setStyleSheetBaseHref;
  }
});
Object.defineProperty(exports, "domtoimage", {
  enumerable: true,
  get: function get() {
    return _domToImage["default"];
  }
});
Object.defineProperty(exports, "getFrequency", {
  enumerable: true,
  get: function get() {
    return _aggregateUtils.getFrequency;
  }
});
Object.defineProperty(exports, "getMode", {
  enumerable: true,
  get: function get() {
    return _aggregateUtils.getMode;
  }
});
Object.defineProperty(exports, "aggregate", {
  enumerable: true,
  get: function get() {
    return _aggregateUtils.aggregate;
  }
});
Object.defineProperty(exports, "getTimelineFromAnimationConfig", {
  enumerable: true,
  get: function get() {
    return _time.getTimelineFromAnimationConfig;
  }
});
Object.defineProperty(exports, "getTimelineFromFilter", {
  enumerable: true,
  get: function get() {
    return _time.getTimelineFromFilter;
  }
});
Object.defineProperty(exports, "SAMPLE_TIMELINE", {
  enumerable: true,
  get: function get() {
    return _time.SAMPLE_TIMELINE;
  }
});
Object.defineProperty(exports, "TIMELINE_MODES", {
  enumerable: true,
  get: function get() {
    return _time.TIMELINE_MODES;
  }
});
Object.defineProperty(exports, "datasetColorMaker", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.datasetColorMaker;
  }
});
Object.defineProperty(exports, "findDefaultColorField", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.findDefaultColorField;
  }
});
Object.defineProperty(exports, "ACCEPTED_ANALYZER_TYPES", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.ACCEPTED_ANALYZER_TYPES;
  }
});
Object.defineProperty(exports, "validateInputData", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.validateInputData;
  }
});
Object.defineProperty(exports, "getSampleForTypeAnalyze", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.getSampleForTypeAnalyze;
  }
});
Object.defineProperty(exports, "getFieldsFromData", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.getFieldsFromData;
  }
});
Object.defineProperty(exports, "renameDuplicateFields", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.renameDuplicateFields;
  }
});
Object.defineProperty(exports, "analyzerTypeToFieldType", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.analyzerTypeToFieldType;
  }
});
Object.defineProperty(exports, "getFormatLabels", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.getFormatLabels;
  }
});
Object.defineProperty(exports, "getFieldFormatLabels", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.getFieldFormatLabels;
  }
});
Object.defineProperty(exports, "getFormatValue", {
  enumerable: true,
  get: function get() {
    return _format.getFormatValue;
  }
});
Object.defineProperty(exports, "exportMapToHTML", {
  enumerable: true,
  get: function get() {
    return _exportMapHtml.exportMapToHTML;
  }
});
Object.defineProperty(exports, "DEFAULT_IMAGE_NAME", {
  enumerable: true,
  get: function get() {
    return _exportUtils.DEFAULT_IMAGE_NAME;
  }
});
Object.defineProperty(exports, "DEFAULT_HTML_NAME", {
  enumerable: true,
  get: function get() {
    return _exportUtils.DEFAULT_HTML_NAME;
  }
});
Object.defineProperty(exports, "DEFAULT_JSON_NAME", {
  enumerable: true,
  get: function get() {
    return _exportUtils.DEFAULT_JSON_NAME;
  }
});
Object.defineProperty(exports, "DEFAULT_DATA_NAME", {
  enumerable: true,
  get: function get() {
    return _exportUtils.DEFAULT_DATA_NAME;
  }
});
Object.defineProperty(exports, "DEFAULT_EXPORT_JSON_SETTINGS", {
  enumerable: true,
  get: function get() {
    return _exportUtils.DEFAULT_EXPORT_JSON_SETTINGS;
  }
});
Object.defineProperty(exports, "isMSEdge", {
  enumerable: true,
  get: function get() {
    return _exportUtils.isMSEdge;
  }
});
Object.defineProperty(exports, "getScaleFromImageSize", {
  enumerable: true,
  get: function get() {
    return _exportUtils.getScaleFromImageSize;
  }
});
Object.defineProperty(exports, "calculateExportImageSize", {
  enumerable: true,
  get: function get() {
    return _exportUtils.calculateExportImageSize;
  }
});
Object.defineProperty(exports, "convertToPng", {
  enumerable: true,
  get: function get() {
    return _exportUtils.convertToPng;
  }
});
Object.defineProperty(exports, "dataURItoBlob", {
  enumerable: true,
  get: function get() {
    return _exportUtils.dataURItoBlob;
  }
});
Object.defineProperty(exports, "downloadFile", {
  enumerable: true,
  get: function get() {
    return _exportUtils.downloadFile;
  }
});
Object.defineProperty(exports, "exportImage", {
  enumerable: true,
  get: function get() {
    return _exportUtils.exportImage;
  }
});
Object.defineProperty(exports, "exportToJsonString", {
  enumerable: true,
  get: function get() {
    return _exportUtils.exportToJsonString;
  }
});
Object.defineProperty(exports, "getMapJSON", {
  enumerable: true,
  get: function get() {
    return _exportUtils.getMapJSON;
  }
});
Object.defineProperty(exports, "exportJson", {
  enumerable: true,
  get: function get() {
    return _exportUtils.exportJson;
  }
});
Object.defineProperty(exports, "exportHtml", {
  enumerable: true,
  get: function get() {
    return _exportUtils.exportHtml;
  }
});
Object.defineProperty(exports, "exportMap", {
  enumerable: true,
  get: function get() {
    return _exportUtils.exportMap;
  }
});
Object.defineProperty(exports, "exporters", {
  enumerable: true,
  get: function get() {
    return _exportUtils["default"];
  }
});
Object.defineProperty(exports, "setLayerBlending", {
  enumerable: true,
  get: function get() {
    return _glUtils.setLayerBlending;
  }
});
Object.defineProperty(exports, "flattenMessages", {
  enumerable: true,
  get: function get() {
    return _localeUtils.flattenMessages;
  }
});
Object.defineProperty(exports, "mergeMessages", {
  enumerable: true,
  get: function get() {
    return _localeUtils.mergeMessages;
  }
});
Object.defineProperty(exports, "validateToken", {
  enumerable: true,
  get: function get() {
    return _mapboxUtils.validateToken;
  }
});
Object.defineProperty(exports, "getDefaultLayerGroupVisibility", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.getDefaultLayerGroupVisibility;
  }
});
Object.defineProperty(exports, "editTopMapStyle", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.editTopMapStyle;
  }
});
Object.defineProperty(exports, "editBottomMapStyle", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.editBottomMapStyle;
  }
});
Object.defineProperty(exports, "isValidStyleUrl", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.isValidStyleUrl;
  }
});
Object.defineProperty(exports, "getStyleDownloadUrl", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.getStyleDownloadUrl;
  }
});
Object.defineProperty(exports, "getStyleImageIcon", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.getStyleImageIcon;
  }
});
Object.defineProperty(exports, "scaleMapStyleByResolution", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.scaleMapStyleByResolution;
  }
});
Object.defineProperty(exports, "mergeLayerGroupVisibility", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.mergeLayerGroupVisibility;
  }
});
Object.defineProperty(exports, "isValidMapInfo", {
  enumerable: true,
  get: function get() {
    return _mapInfoUtils.isValidMapInfo;
  }
});
Object.defineProperty(exports, "snapToMarks", {
  enumerable: true,
  get: function get() {
    return _plot.snapToMarks;
  }
});
Object.defineProperty(exports, "computeDeckEffects", {
  enumerable: true,
  get: function get() {
    return _effectUtils.computeDeckEffects;
  }
});
Object.defineProperty(exports, "fixEffectOrder", {
  enumerable: true,
  get: function get() {
    return _effectUtils.fixEffectOrder;
  }
});
Object.defineProperty(exports, "reorderEffectOrder", {
  enumerable: true,
  get: function get() {
    return _effectUtils.reorderEffectOrder;
  }
});
Object.defineProperty(exports, "validateEffectParameters", {
  enumerable: true,
  get: function get() {
    return _effectUtils.validateEffectParameters;
  }
});
Object.defineProperty(exports, "transformRequest", {
  enumerable: true,
  get: function get() {
    return _mapboxUtils2.transformRequest;
  }
});
Object.defineProperty(exports, "isStyleUsingMapboxTiles", {
  enumerable: true,
  get: function get() {
    return _mapboxUtils2.isStyleUsingMapboxTiles;
  }
});
Object.defineProperty(exports, "createDataContainer", {
  enumerable: true,
  get: function get() {
    return _dataContainerUtils.createDataContainer;
  }
});
Object.defineProperty(exports, "createIndexedDataContainer", {
  enumerable: true,
  get: function get() {
    return _dataContainerUtils.createIndexedDataContainer;
  }
});
Object.defineProperty(exports, "getSampleContainerData", {
  enumerable: true,
  get: function get() {
    return _dataContainerUtils.getSampleData;
  }
});
Object.defineProperty(exports, "DataForm", {
  enumerable: true,
  get: function get() {
    return _dataContainerUtils.DataForm;
  }
});
Object.defineProperty(exports, "ArrowDataContainer", {
  enumerable: true,
  get: function get() {
    return _arrowDataContainer.ArrowDataContainer;
  }
});
Object.defineProperty(exports, "arrowDataTypeToFieldType", {
  enumerable: true,
  get: function get() {
    return _arrowDataContainer.arrowDataTypeToFieldType;
  }
});
Object.defineProperty(exports, "arrowDataTypeToAnalyzerDataType", {
  enumerable: true,
  get: function get() {
    return _arrowDataContainer.arrowDataTypeToAnalyzerDataType;
  }
});
Object.defineProperty(exports, "getQuantileDomain", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getQuantileDomain;
  }
});
Object.defineProperty(exports, "getOrdinalDomain", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getOrdinalDomain;
  }
});
Object.defineProperty(exports, "getLinearDomain", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getLinearDomain;
  }
});
Object.defineProperty(exports, "getLogDomain", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getLogDomain;
  }
});
Object.defineProperty(exports, "DataRow", {
  enumerable: true,
  get: function get() {
    return _dataRow.DataRow;
  }
});
Object.defineProperty(exports, "getCentroid", {
  enumerable: true,
  get: function get() {
    return _h3Utils.getCentroid;
  }
});
Object.defineProperty(exports, "idToPolygonGeo", {
  enumerable: true,
  get: function get() {
    return _h3Utils.idToPolygonGeo;
  }
});
Object.defineProperty(exports, "h3IsValid", {
  enumerable: true,
  get: function get() {
    return _h3Utils.h3IsValid;
  }
});
Object.defineProperty(exports, "getHexFields", {
  enumerable: true,
  get: function get() {
    return _h3Utils.getHexFields;
  }
});

var _colorUtils = require("./color-utils");

var _notificationsUtils = require("./notifications-utils");

var _domUtils = require("./dom-utils");

var _domToImage = _interopRequireDefault(require("./dom-to-image"));

var _aggregateUtils = require("./aggregate-utils");

var _dataUtils = require("./data-utils");

Object.keys(_dataUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dataUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _dataUtils[key];
    }
  });
});

var _time = require("./time");

var _datasetUtils = require("./dataset-utils");

var _format = require("./format");

var _exportMapHtml = require("./export-map-html");

var _exportUtils = _interopRequireWildcard(require("./export-utils"));

var _glUtils = require("./gl-utils");

var _localeUtils = require("./locale-utils");

var _observeDimensions = require("./observe-dimensions");

Object.keys(_observeDimensions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _observeDimensions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _observeDimensions[key];
    }
  });
});

var _projectionUtils = require("./projection-utils");

Object.keys(_projectionUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _projectionUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _projectionUtils[key];
    }
  });
});

var _mapboxUtils = require("./mapbox-utils");

var _mapboxGlStyleEditor = require("./map-style-utils/mapbox-gl-style-editor");

var _mapInfoUtils = require("./map-info-utils");

var _utils = require("./utils");

Object.keys(_utils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _utils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _utils[key];
    }
  });
});

var _splitMapUtils = require("./split-map-utils");

Object.keys(_splitMapUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _splitMapUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _splitMapUtils[key];
    }
  });
});

var _plot = require("./plot");

var _effectUtils = require("./effect-utils");

var _mapboxUtils2 = require("./map-style-utils/mapbox-utils");

var _mapUtils = require("./map-utils");

Object.keys(_mapUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mapUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _mapUtils[key];
    }
  });
});

var _dataContainerUtils = require("./data-container-utils");

var _arrowDataContainer = require("./arrow-data-container");

var _filterUtils = require("./filter-utils");

Object.keys(_filterUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _filterUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _filterUtils[key];
    }
  });
});

var _dataScaleUtils = require("./data-scale-utils");

var _dataRow = require("./data-row");

var _h3Utils = require("./h3-utils");
