// kepler.gl/src/utils/dist/time.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getTimelineFromFilter = exports.getTimelineFromAnimationConfig = exports.SAMPLE_TIMELINE = exports.TIMELINE_MODES = void 0;

var _utils = require("./utils");

// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var TIMELINE_MODES = {
  inner: 'inner',
  outer: 'outer'
};
exports.TIMELINE_MODES = TIMELINE_MODES;
var SAMPLE_TIMELINE = {
  // value: [15], // represent 15% of the all width
  value: [5, 15],
  // represent start at 5% and ends at 15%
  domain: [1, 100],
  // represent the total domain
  speed: 1,
  enableInteraction: false,
  // can use interact with this timeline
  isAnimating: false,
  step: null,
  // @todo: giuseppe coverType: 'inner' | 'outer'
  mode: TIMELINE_MODES.inner //  ....

};
exports.SAMPLE_TIMELINE = SAMPLE_TIMELINE;

var getTimelineFromAnimationConfig = function getTimelineFromAnimationConfig(animationConfig) {
  var currentTime = animationConfig.currentTime,
      domain = animationConfig.domain,
      speed = animationConfig.speed,
      isAnimating = animationConfig.isAnimating,
      timeSteps = animationConfig.timeSteps,
      defaultTimeFormat = animationConfig.defaultTimeFormat,
      timeFormat = animationConfig.timeFormat,
      timezone = animationConfig.timezone;
  return {
    // @ts-expect-error
    value: (0, _utils.toArray)(currentTime),
    enableInteraction: true,
    // @ts-expect-error
    domain: domain,
    speed: speed,
    isAnimating: isAnimating || false,
    timeSteps: timeSteps,
    defaultTimeFormat: defaultTimeFormat,
    timeFormat: timeFormat,
    timezone: timezone,
    marks: null
  };
};

exports.getTimelineFromAnimationConfig = getTimelineFromAnimationConfig;

var getTimelineFromFilter = function getTimelineFromFilter(filter) {
  var value = filter.value,
      domain = filter.domain,
      speed = filter.speed,
      isAnimating = filter.isAnimating,
      step = filter.step,
      timeSteps = filter.timeSteps,
      defaultTimeFormat = filter.defaultTimeFormat,
      timeFormat = filter.timeFormat,
      timezone = filter.timezone,
      animationWindow = filter.animationWindow;
  return {
    value: value,
    enableInteraction: true,
    domain: domain,
    speed: speed,
    isAnimating: isAnimating,
    step: step,
    timeSteps: timeSteps,
    defaultTimeFormat: defaultTimeFormat,
    timeFormat: timeFormat,
    timezone: timezone,
    animationWindow: animationWindow,
    marks: null
  };
};

exports.getTimelineFromFilter = getTimelineFromFilter;
