// kepler.gl/src/layers/dist/layer-utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assignPointPairToLayerColumn = assignPointPairToLayerColumn;
exports.getGeojsonLayerMetaFromArrow = getGeojsonLayerMetaFromArrow;
exports.isLayerHoveredFromArrow = isLayerHoveredFromArrow;
exports.getHoveredObjectFromArrow = getHoveredObjectFromArrow;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _arrow = require("@loaders.gl/arrow");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function assignPointPairToLayerColumn(pair, hasAlt) {
  var _pair$pair = pair.pair,
      lat = _pair$pair.lat,
      lng = _pair$pair.lng,
      alt = _pair$pair.alt;

  if (!hasAlt) {
    return {
      lat: lat,
      lng: lng
    };
  }

  var defaultAltColumn = {
    value: null,
    fieldIdx: -1,
    optional: true
  };
  return {
    lat: lat,
    lng: lng,
    altitude: alt ? _objectSpread(_objectSpread({}, defaultAltColumn), alt) : defaultAltColumn
  };
}

function getGeojsonLayerMetaFromArrow(_ref) {
  var _arrowField$metadata;

  var dataContainer = _ref.dataContainer,
      getGeoColumn = _ref.getGeoColumn,
      getGeoField = _ref.getGeoField,
      chunkIndex = _ref.chunkIndex;
  var geoColumn = getGeoColumn(dataContainer);
  var arrowField = getGeoField(dataContainer);
  var encoding = arrowField === null || arrowField === void 0 ? void 0 : (_arrowField$metadata = arrowField.metadata) === null || _arrowField$metadata === void 0 ? void 0 : _arrowField$metadata.get('ARROW:extension:name');

  var options = _objectSpread(_objectSpread({}, chunkIndex !== undefined && chunkIndex >= 0 ? {
    chunkIndex: chunkIndex,
    chunkOffset: geoColumn.data[0].length * chunkIndex
  } : {}), {}, {
    triangulate: true,
    calculateMeanCenters: true
  }); // create binary data from arrow data for GeoJsonLayer


  var _getBinaryGeometriesF = (0, _arrow.getBinaryGeometriesFromArrow)(geoColumn, encoding, options),
      binaryGeometries = _getBinaryGeometriesF.binaryGeometries,
      featureTypes = _getBinaryGeometriesF.featureTypes,
      bounds = _getBinaryGeometriesF.bounds,
      meanCenters = _getBinaryGeometriesF.meanCenters; // since there is no feature.properties.radius, we set fixedRadius to false


  var fixedRadius = false;
  return {
    dataToFeature: binaryGeometries,
    featureTypes: featureTypes,
    bounds: bounds,
    fixedRadius: fixedRadius,
    centroids: meanCenters
  };
}

function isLayerHoveredFromArrow(objectInfo, layerId) {
  // there could be multiple deck.gl layers created from multiple chunks in arrow table
  // the objectInfo.layer id should be `${this.id}-${i}`
  if (objectInfo !== null && objectInfo !== void 0 && objectInfo.picked) {
    var _objectInfo$layer, _objectInfo$layer$pro;

    var deckLayerId = objectInfo === null || objectInfo === void 0 ? void 0 : (_objectInfo$layer = objectInfo.layer) === null || _objectInfo$layer === void 0 ? void 0 : (_objectInfo$layer$pro = _objectInfo$layer.props) === null || _objectInfo$layer$pro === void 0 ? void 0 : _objectInfo$layer$pro.id;
    return deckLayerId.startsWith(layerId);
  }

  return false;
}

function getHoveredObjectFromArrow(objectInfo, dataContainer, layerId, columnAccessor, fieldAccessor) {
  // hover object returns the index of the object in the data array
  // NOTE: this could be done in Deck.gl getPickingInfo(params) and binaryToGeojson()
  if (isLayerHoveredFromArrow(objectInfo, layerId) && objectInfo.index >= 0 && dataContainer) {
    var _field$metadata;

    var col = columnAccessor(dataContainer);
    var rawGeometry = col === null || col === void 0 ? void 0 : col.get(objectInfo.index);
    var field = fieldAccessor(dataContainer);
    var encoding = field === null || field === void 0 ? void 0 : (_field$metadata = field.metadata) === null || _field$metadata === void 0 ? void 0 : _field$metadata.get('ARROW:extension:name');
    var hoveredFeature = (0, _arrow.parseGeometryFromArrow)(rawGeometry, encoding);
    var properties = dataContainer.rowAsArray(objectInfo.index).reduce(function (prev, cur, i) {
      var _dataContainer$getFie;

      var fieldName = dataContainer === null || dataContainer === void 0 ? void 0 : (_dataContainer$getFie = dataContainer.getField) === null || _dataContainer$getFie === void 0 ? void 0 : _dataContainer$getFie.call(dataContainer, i).name;

      if (fieldName !== field.name) {
        prev[fieldName] = cur;
      }

      return prev;
    }, {});
    return hoveredFeature ? {
      type: 'Feature',
      geometry: hoveredFeature,
      properties: _objectSpread(_objectSpread({}, properties), {}, {
        index: objectInfo.index
      })
    } : null;
  }

  return null;
}
