// @hubble.gl/core/dist/es5/encoders/video/stream-encoder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _frameEncoder = _interopRequireDefault(require("../frame-encoder"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var StreamEncoder = function (_FrameEncoder) {
  (0, _inherits2.default)(StreamEncoder, _FrameEncoder);

  var _super = _createSuper(StreamEncoder);

  function StreamEncoder(settings) {
    var _this;

    (0, _classCallCheck2.default)(this, StreamEncoder);
    _this = _super.call(this, settings);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "stream", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "mediaRecorder", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "chunks", void 0);
    _this.mimeType = 'video/webm';
    _this.extension = '.webm';
    _this.stream = null;
    _this.mediaRecorder = null;
    _this.chunks = [];
    _this.start = _this.start.bind((0, _assertThisInitialized2.default)(_this));
    _this.add = _this.add.bind((0, _assertThisInitialized2.default)(_this));
    _this.save = _this.save.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(StreamEncoder, [{
    key: "start",
    value: function start() {
      this.stream = null;
      this.mediaRecorder = null;
      this.chunks = [];
    }
  }, {
    key: "add",
    value: function () {
      var _add = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee(canvas) {
        var _this2 = this;

        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this.stream) {
                  this.stream = canvas.captureStream(this.framerate);
                  this.mediaRecorder = new MediaRecorder(this.stream);
                  this.mediaRecorder.start();

                  this.mediaRecorder.ondataavailable = function (e) {
                    _this2.chunks.push(e.data);
                  };
                }

                return _context.abrupt("return", Promise.resolve());

              case 2:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function add(_x) {
        return _add.apply(this, arguments);
      }

      return add;
    }()
  }, {
    key: "save",
    value: function () {
      var _save = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee2() {
        var _this3 = this;

        var waiting;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                waiting = new Promise(function (resolve) {
                  _this3.mediaRecorder.onstop = function () {
                    var blob = new Blob(_this3.chunks, {
                      type: 'video/webm'
                    });
                    _this3.chunks = [];
                    resolve(blob);
                  };

                  _this3.mediaRecorder.stop();
                });
                return _context2.abrupt("return", waiting);

              case 2:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }));

      function save() {
        return _save.apply(this, arguments);
      }

      return save;
    }()
  }]);
  return StreamEncoder;
}(_frameEncoder.default);

var _default = StreamEncoder;
exports.default = _default;
