// @hubble.gl/core/dist/es5/encoders/video/webm-encoder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _webmWriter = _interopRequireDefault(require("webm-writer"));

var _frameEncoder = _interopRequireDefault(require("../frame-encoder"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var WebMEncoder = function (_FrameEncoder) {
  (0, _inherits2.default)(WebMEncoder, _FrameEncoder);

  var _super = _createSuper(WebMEncoder);

  function WebMEncoder(settings) {
    var _this;

    (0, _classCallCheck2.default)(this, WebMEncoder);
    _this = _super.call(this, settings);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "videoWriter", void 0);
    _this.quality = 0.8;

    if (settings.webm && settings.webm.quality) {
      _this.quality = settings.webm.quality;
    }

    var canvas = document.createElement('canvas');

    if (canvas.toDataURL('image/webp').substr(5, 10) !== 'image/webp') {
      console.error('WebP not supported - try another export format');
    }

    _this.extension = '.webm';
    _this.mimeType = 'video/webm';
    _this.videoWriter = null;
    _this.start = _this.start.bind((0, _assertThisInitialized2.default)(_this));
    _this.add = _this.add.bind((0, _assertThisInitialized2.default)(_this));
    _this.save = _this.save.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(WebMEncoder, [{
    key: "start",
    value: function start() {
      this.videoWriter = new _webmWriter.default({
        quality: this.quality,
        fileWriter: null,
        fd: null,
        frameRate: this.framerate
      });
    }
  }, {
    key: "add",
    value: function () {
      var _add = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee(canvas) {
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                this.videoWriter.addFrame(canvas);
                _context.next = 3;
                return Promise.resolve();

              case 3:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function add(_x) {
        return _add.apply(this, arguments);
      }

      return add;
    }()
  }, {
    key: "save",
    value: function () {
      var _save = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee2() {
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                return _context2.abrupt("return", this.videoWriter.complete());

              case 1:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function save() {
        return _save.apply(this, arguments);
      }

      return save;
    }()
  }]);
  return WebMEncoder;
}(_frameEncoder.default);

var _default = WebMEncoder;
exports.default = _default;
