// @kepler.gl/components/dist/common/data-table/grid.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactVirtualized = require("react-virtualized");

var _lodash = _interopRequireDefault(require("lodash.isequal"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var GridHack = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(GridHack, _PureComponent);

  var _super = _createSuper(GridHack);

  function GridHack() {
    var _this;

    (0, _classCallCheck2["default"])(this, GridHack);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "grid", null);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_preventScrollBack", function (e) {
      var scrollLeft = _this.props.scrollLeft;

      if (scrollLeft !== undefined && scrollLeft <= 0 && e.deltaX < 0) {
        // Prevent Scroll On Scrollable Elements, avoid browser backward navigation
        // https://alvarotrigo.com/blog/prevent-scroll-on-scrollable-element-js/
        e.preventDefault();
        e.stopPropagation();
        return false;
      }

      return;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_updateRef", function (x) {
      if (!_this.grid && x) {
        var _this$grid, _this$grid$_scrolling;

        _this.grid = x;
        /*
         * This hack exists because we need to add wheel event listener to the div rendered by Grid
         *
         */
        // @ts-expect-error _scrollingContainer not typed in Grid

        (_this$grid = _this.grid) === null || _this$grid === void 0 ? void 0 : (_this$grid$_scrolling = _this$grid._scrollingContainer) === null || _this$grid$_scrolling === void 0 ? void 0 : _this$grid$_scrolling.addEventListener('wheel', _this._preventScrollBack, {
          passive: false
        });
      }
    });
    return _this;
  }

  (0, _createClass2["default"])(GridHack, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(preProps) {
      /*
       * This hack exists because in react-virtualized the
       * _columnWidthGetter is only called in the constructor
       * even though it is reassigned with new props resulting in
       * a new width for cells not being calculated so we must
       * force trigger a resize.
       *
       * https://github.com/bvaughn/react-virtualized/blob/master/source/Grid/Grid.js#L322
       *
       */
      if (!(0, _lodash["default"])(preProps.cellSizeCache, this.props.cellSizeCache)) {
        var _this$grid2;

        (_this$grid2 = this.grid) === null || _this$grid2 === void 0 ? void 0 : _this$grid2.recomputeGridSize();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var _this$grid3, _this$grid3$_scrollin;

      // @ts-expect-error _scrollingContainer not typed in Grid
      (_this$grid3 = this.grid) === null || _this$grid3 === void 0 ? void 0 : (_this$grid3$_scrollin = _this$grid3._scrollingContainer) === null || _this$grid3$_scrollin === void 0 ? void 0 : _this$grid3$_scrollin.removeEventListener('wheel', this._preventScrollBack, {
        passive: false
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          setGridRef = _this$props.setGridRef,
          rest = (0, _objectWithoutProperties2["default"])(_this$props, ["setGridRef"]);
      return /*#__PURE__*/_react["default"].createElement(_reactVirtualized.Grid, (0, _extends2["default"])({
        ref: function ref(x) {
          if (setGridRef) setGridRef(x);

          _this2._updateRef(x);
        },
        key: "grid-hack"
      }, rest));
    }
  }]);
  return GridHack;
}(_react.PureComponent);

exports["default"] = GridHack;
