// @kepler.gl/components/dist/modals/cloud-components/cloud-header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CloudHeader = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = require("../../common/styled-components");

var _icons = require("../../common/icons");

var _localization = require("@kepler.gl/localization");

var _styledComponents2 = _interopRequireDefault(require("styled-components"));

var _constants = require("@kepler.gl/constants");

var _templateObject, _templateObject2, _templateObject3;

var StyledStorageHeader = _styledComponents2["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: row;\n  justify-content: space-between;\n  align-items: center;\n  margin-bottom: 16px;\n  font-size: 12px;\n  line-height: 14px;\n"])));

var StyledBackBtn = _styledComponents2["default"].a(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  margin-bottom: 16px;\n  color: #3a414c;\n  cursor: pointer;\n\n  &:hover {\n    font-weight: 500;\n  }\n"])));

var LINK_STYLE = {
  textDecoration: 'underline'
};

var Title = _styledComponents2["default"].span(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 14px;\n  line-height: 16px;\n  font-weight: 500;\n  margin-bottom: 16px;\n\n  span {\n    text-transform: capitalize;\n  }\n"])));

var CloudHeader = function CloudHeader(_ref) {
  var provider = _ref.provider,
      onBack = _ref.onBack;
  var managementUrl = (0, _react.useMemo)(function () {
    return provider.getManagementUrl();
  }, [provider]);
  return /*#__PURE__*/_react["default"].createElement("div", {
    "data-testid": _constants.dataTestIds.cloudHeader
  }, /*#__PURE__*/_react["default"].createElement(StyledStorageHeader, null, /*#__PURE__*/_react["default"].createElement(StyledBackBtn, null, /*#__PURE__*/_react["default"].createElement(_styledComponents.Button, {
    link: true,
    onClick: onBack
  }, /*#__PURE__*/_react["default"].createElement(_icons.ArrowLeft, {
    height: "14px"
  }), /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
    id: 'modal.loadStorageMap.back'
  }))), managementUrl && /*#__PURE__*/_react["default"].createElement("a", {
    key: 1,
    href: managementUrl,
    target: "_blank",
    rel: "noopener noreferrer",
    style: LINK_STYLE
  }, provider.displayName)), /*#__PURE__*/_react["default"].createElement(Title, null, /*#__PURE__*/_react["default"].createElement("span", null, provider.displayName), /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
    id: 'modal.loadStorageMap.storageMaps'
  })));
};

exports.CloudHeader = CloudHeader;
