// @hubble.gl/core/dist/esm/scene/deck-scene.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";

var DeckScene = function () {
  function DeckScene(_ref) {
    var animationLoop = _ref.animationLoop,
        keyframes = _ref.keyframes,
        data = _ref.data,
        _ref$renderLayers = _ref.renderLayers,
        renderLayers = _ref$renderLayers === void 0 ? undefined : _ref$renderLayers,
        lengthMs = _ref.lengthMs,
        width = _ref.width,
        height = _ref.height,
        _ref$currentCamera = _ref.currentCamera,
        currentCamera = _ref$currentCamera === void 0 ? undefined : _ref$currentCamera;

    _classCallCheck(this, DeckScene);

    this.animationLoop = animationLoop;
    this.keyframes = keyframes;
    this.data = data;
    this._renderLayers = renderLayers;
    this.renderLayers = this.renderLayers.bind(this);
    this.lengthMs = lengthMs;
    this.width = width;
    this.height = height;
    this.currentCamera = currentCamera;
  }

  _createClass(DeckScene, [{
    key: "hasLayers",
    value: function hasLayers() {
      return Boolean(this._renderLayers);
    }
  }, {
    key: "renderLayers",
    value: function renderLayers() {
      return this._renderLayers(this);
    }
  }]);

  return DeckScene;
}();

export { DeckScene as default };
