// @hubble.gl/core/dist/esm/scene/kepler-scene.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";

var KeplerScene = function () {
  function KeplerScene(_ref) {
    var animationLoop = _ref.animationLoop,
        keyframes = _ref.keyframes,
        data = _ref.data,
        filters = _ref.filters,
        getFrame = _ref.getFrame,
        lengthMs = _ref.lengthMs,
        width = _ref.width,
        height = _ref.height;

    _classCallCheck(this, KeplerScene);

    this.animationLoop = animationLoop;
    this.keyframes = keyframes;
    this.data = data;
    this._getFrame = getFrame;
    this.filters = filters;
    this.getFrame = this.getFrame.bind(this);
    this.lengthMs = lengthMs;
    this.width = width;
    this.height = height;
  }

  _createClass(KeplerScene, [{
    key: "getFrame",
    value: function getFrame(keplerGl) {
      return this._getFrame(keplerGl, this.keyframes, this.filters);
    }
  }]);

  return KeplerScene;
}();

export { KeplerScene as default };
