// @hubble.gl/core/dist/es5/keyframes/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Keyframes", {
  enumerable: true,
  get: function get() {
    return _keyframes["default"];
  }
});
Object.defineProperty(exports, "CameraKeyframes", {
  enumerable: true,
  get: function get() {
    return _cameraKeyframes["default"];
  }
});
Object.defineProperty(exports, "hold", {
  enumerable: true,
  get: function get() {
    return _easings.hold;
  }
});
Object.defineProperty(exports, "FilterValueKeyframes", {
  enumerable: true,
  get: function get() {
    return _kepler.FilterValueKeyframes;
  }
});
Object.defineProperty(exports, "LayerKeyframes", {
  enumerable: true,
  get: function get() {
    return _layers.LayerKeyframes;
  }
});
Object.defineProperty(exports, "GridLayerKeyframes", {
  enumerable: true,
  get: function get() {
    return _layers.GridLayerKeyframes;
  }
});
Object.defineProperty(exports, "ScatterPlotLayerKeyframes", {
  enumerable: true,
  get: function get() {
    return _layers.ScatterPlotLayerKeyframes;
  }
});

var _keyframes = _interopRequireDefault(require("./keyframes"));

var _cameraKeyframes = _interopRequireDefault(require("./camera-keyframes"));

var _easings = require("./easings");

var _kepler = require("./kepler");

var _layers = require("./layers");
