// @hubble.gl/core/dist/es5/scene/deck-scene.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var DeckScene = function () {
  function DeckScene(_ref) {
    var animationLoop = _ref.animationLoop,
        keyframes = _ref.keyframes,
        data = _ref.data,
        _ref$renderLayers = _ref.renderLayers,
        renderLayers = _ref$renderLayers === void 0 ? undefined : _ref$renderLayers,
        lengthMs = _ref.lengthMs,
        width = _ref.width,
        height = _ref.height,
        _ref$currentCamera = _ref.currentCamera,
        currentCamera = _ref$currentCamera === void 0 ? undefined : _ref$currentCamera;
    (0, _classCallCheck2["default"])(this, DeckScene);
    this.animationLoop = animationLoop;
    this.keyframes = keyframes;
    this.data = data;
    this._renderLayers = renderLayers;
    this.renderLayers = this.renderLayers.bind(this);
    this.lengthMs = lengthMs;
    this.width = width;
    this.height = height;
    this.currentCamera = currentCamera;
  }

  (0, _createClass2["default"])(DeckScene, [{
    key: "hasLayers",
    value: function hasLayers() {
      return Boolean(this._renderLayers);
    }
  }, {
    key: "renderLayers",
    value: function renderLayers() {
      return this._renderLayers(this);
    }
  }]);
  return DeckScene;
}();

exports["default"] = DeckScene;
