// @hubble.gl/core/dist/es5/scene/kepler-scene.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var KeplerScene = function () {
  function KeplerScene(_ref) {
    var animationLoop = _ref.animationLoop,
        keyframes = _ref.keyframes,
        data = _ref.data,
        filters = _ref.filters,
        getFrame = _ref.getFrame,
        lengthMs = _ref.lengthMs,
        width = _ref.width,
        height = _ref.height;
    (0, _classCallCheck2["default"])(this, KeplerScene);
    this.animationLoop = animationLoop;
    this.keyframes = keyframes;
    this.data = data;
    this._getFrame = getFrame;
    this.filters = filters;
    this.getFrame = this.getFrame.bind(this);
    this.lengthMs = lengthMs;
    this.width = width;
    this.height = height;
  }

  (0, _createClass2["default"])(KeplerScene, [{
    key: "getFrame",
    value: function getFrame(keplerGl) {
      return this._getFrame(keplerGl, this.keyframes, this.filters);
    }
  }]);
  return KeplerScene;
}();

exports["default"] = KeplerScene;
