// kepler.gl/src/components/dist/modals/cloud-components/cloud-maps.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CloudMaps = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _loadingDialog = _interopRequireDefault(require("../loading-dialog"));

var _localization = require("@kepler.gl/localization");

var _cloudItem = require("./cloud-item");

var _flexContainer = require("../../common/flex-container");

var _templateObject, _templateObject2;

var StyledSpinner = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  text-align: center;\n  span {\n    margin: 0 auto;\n  }\n"])));

var StyledFlexContainer = (0, _styledComponents["default"])(_flexContainer.FlexContainer)(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  justify-content: flex-start;\n  flex-wrap: wrap;\n"])));

var CloudMaps = function CloudMaps(_ref) {
  var provider = _ref.provider,
      onSelectMap = _ref.onSelectMap,
      isLoading = _ref.isLoading,
      maps = _ref.maps,
      error = _ref.error;

  if (error) {
    return /*#__PURE__*/_react["default"].createElement("div", null, "Error while fetching maps: ", error.message);
  }

  if (isLoading) {
    return /*#__PURE__*/_react["default"].createElement(StyledSpinner, null, /*#__PURE__*/_react["default"].createElement(_loadingDialog["default"], {
      size: 64
    }));
  }

  return /*#__PURE__*/_react["default"].createElement(StyledFlexContainer, null, (maps !== null && maps !== void 0 ? maps : []).length ? maps.map(function (vis) {
    return /*#__PURE__*/_react["default"].createElement(_cloudItem.CloudItem, {
      key: vis.id,
      onClick: function onClick() {
        return onSelectMap(provider, vis);
      },
      vis: vis
    });
  }) : /*#__PURE__*/_react["default"].createElement("div", {
    className: "visualization-list__message"
  }, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
    id: 'modal.loadStorageMap.noSavedMaps'
  })));
};

exports.CloudMaps = CloudMaps;
