// kepler.gl/src/components/dist/modals/error-display.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _errorBoundary = _interopRequireDefault(require("../common/error-boundary"));

var _notificationItem = _interopRequireDefault(require("../notification-panel/notification-item"));

// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var NotificationItem = (0, _notificationItem["default"])();

var ErrorDisplay = function ErrorDisplay(_ref) {
  var error = _ref.error;
  var notification = (0, _react.useMemo)(function () {
    return {
      type: 'error',
      message: error,
      id: 'cloud-export-error'
    };
  }, [error]);
  return /*#__PURE__*/_react["default"].createElement(_errorBoundary["default"], null, /*#__PURE__*/_react["default"].createElement(NotificationItem, {
    notification: notification,
    isExpanded: true
  }));
};

var _default = ErrorDisplay;
exports["default"] = _default;
