// @kepler.gl/components/dist/context.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.RootContext = exports.CloudProviderContext = exports.FeatureFlagsContextProvider = exports.FeatureFlagsContext = void 0;

var _react = _interopRequireWildcard(require("react"));

// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var identity = function identity(state) {
  return state;
}; // New Context API only supported after 16.3


var KeplerGlContext = /*#__PURE__*/(0, _react.createContext)({
  selector: identity,
  id: 'map'
}); // TODO: breakdown this file into multiple files

var FeatureFlagsContext = /*#__PURE__*/(0, _react.createContext)({});
exports.FeatureFlagsContext = FeatureFlagsContext;

var FeatureFlagsContextProvider = function FeatureFlagsContextProvider(props) {
  return /*#__PURE__*/_react["default"].createElement(FeatureFlagsContext.Provider, {
    value: props.featureFlags || null
  }, props.children);
};
/**
 * This provides keeps track of the ist cloud providers
 * and the current selected one
 */


exports.FeatureFlagsContextProvider = FeatureFlagsContextProvider;
var CloudProviderContext = /*#__PURE__*/(0, _react.createContext)({
  provider: null,
  setProvider: function setProvider() {},
  cloudProviders: []
});
exports.CloudProviderContext = CloudProviderContext;
var RootContext = /*#__PURE__*/(0, _react.createContext)(null);
exports.RootContext = RootContext;
var _default = KeplerGlContext;
exports["default"] = _default;
