// @kepler.gl/components/dist/layer-animation-controller.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _constants = require("@kepler.gl/constants");

var _utils = require("@kepler.gl/utils");

var _animationController = _interopRequireDefault(require("./common/animation-control/animation-controller"));

// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
LayerAnimationControllerFactory.deps = [_animationController["default"]];

function LayerAnimationControllerFactory(AnimationController) {
  var LayerAnimationController = function LayerAnimationController(_ref) {
    var animationConfig = _ref.animationConfig,
        setLayerAnimationTime = _ref.setLayerAnimationTime,
        children = _ref.children;
    var timeSteps = animationConfig.timeSteps,
        domain = animationConfig.domain;
    var setTimelineValue = (0, _react.useCallback)(function (value) {
      if (Array.isArray(timeSteps)) {
        setLayerAnimationTime((0, _utils.snapToMarks)((0, _utils.toArray)(value)[0], timeSteps)); // TODO: merge slider in to avoid this step
      } else if (domain && value >= domain[0] && value <= domain[1]) {
        setLayerAnimationTime(value);
      }
    }, [domain, setLayerAnimationTime, timeSteps]);
    var timeline = (0, _utils.getTimelineFromAnimationConfig)(animationConfig);
    return /*#__PURE__*/_react["default"].createElement(AnimationController, {
      key: "layer-control",
      value: Number(animationConfig.currentTime),
      domain: animationConfig.domain,
      speed: animationConfig.speed,
      isAnimating: animationConfig.isAnimating,
      steps: animationConfig.timeSteps,
      animationWindow: animationConfig.timeSteps ? _constants.ANIMATION_WINDOW.interval : _constants.ANIMATION_WINDOW.point,
      setTimelineValue: setTimelineValue,
      timeline: timeline,
      children: children
    });
  };

  return LayerAnimationController;
}

var _default = LayerAnimationControllerFactory;
exports["default"] = _default;
