// @hubble.gl/core/dist/es5/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Animation", {
  enumerable: true,
  get: function get() {
    return _animations.Animation;
  }
});
Object.defineProperty(exports, "AnimationManager", {
  enumerable: true,
  get: function get() {
    return _animations.AnimationManager;
  }
});
Object.defineProperty(exports, "CameraKeyframes", {
  enumerable: true,
  get: function get() {
    return _keyframes.CameraKeyframes;
  }
});
Object.defineProperty(exports, "DeckAdapter", {
  enumerable: true,
  get: function get() {
    return _adapters.DeckAdapter;
  }
});
Object.defineProperty(exports, "DeckAnimation", {
  enumerable: true,
  get: function get() {
    return _animations.DeckAnimation;
  }
});
Object.defineProperty(exports, "DeckLayerKeyframes", {
  enumerable: true,
  get: function get() {
    return _keyframes.DeckLayerKeyframes;
  }
});
Object.defineProperty(exports, "FrameEncoder", {
  enumerable: true,
  get: function get() {
    return _encoders.FrameEncoder;
  }
});
Object.defineProperty(exports, "GifEncoder", {
  enumerable: true,
  get: function get() {
    return _encoders.GifEncoder;
  }
});
Object.defineProperty(exports, "JPEGEncoder", {
  enumerable: true,
  get: function get() {
    return _encoders.JPEGEncoder;
  }
});
Object.defineProperty(exports, "JPEGSequenceEncoder", {
  enumerable: true,
  get: function get() {
    return _encoders.JPEGSequenceEncoder;
  }
});
Object.defineProperty(exports, "KeplerAnimation", {
  enumerable: true,
  get: function get() {
    return _animations.KeplerAnimation;
  }
});
Object.defineProperty(exports, "KeplerFilterKeyframes", {
  enumerable: true,
  get: function get() {
    return _keyframes.KeplerFilterKeyframes;
  }
});
Object.defineProperty(exports, "KeplerLayerKeyframes", {
  enumerable: true,
  get: function get() {
    return _keyframes.KeplerLayerKeyframes;
  }
});
Object.defineProperty(exports, "KeplerTripKeyframes", {
  enumerable: true,
  get: function get() {
    return _keyframes.KeplerTripKeyframes;
  }
});
Object.defineProperty(exports, "Keyframes", {
  enumerable: true,
  get: function get() {
    return _keyframes.Keyframes;
  }
});
Object.defineProperty(exports, "PNGEncoder", {
  enumerable: true,
  get: function get() {
    return _encoders.PNGEncoder;
  }
});
Object.defineProperty(exports, "PNGSequenceEncoder", {
  enumerable: true,
  get: function get() {
    return _encoders.PNGSequenceEncoder;
  }
});
Object.defineProperty(exports, "PreviewEncoder", {
  enumerable: true,
  get: function get() {
    return _encoders.PreviewEncoder;
  }
});
Object.defineProperty(exports, "WebMEncoder", {
  enumerable: true,
  get: function get() {
    return _encoders.WebMEncoder;
  }
});
Object.defineProperty(exports, "hold", {
  enumerable: true,
  get: function get() {
    return _keyframes.hold;
  }
});
Object.defineProperty(exports, "linear", {
  enumerable: true,
  get: function get() {
    return _keyframes.linear;
  }
});

require("./lib/init");

var _adapters = require("./adapters");

var _encoders = require("./encoders");

var _keyframes = require("./keyframes");

var _animations = require("./animations");
