// @hubble.gl/core/dist/es5/keyframes/kepler-layer-keyframes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _keyframes = _interopRequireDefault(require("./keyframes"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function getFeatures(layer) {
  return Object.keys(layer.config.visConfig);
}

var KeplerLayerKeyframes = function (_Keyframes) {
  (0, _inherits2.default)(KeplerLayerKeyframes, _Keyframes);

  var _super = _createSuper(KeplerLayerKeyframes);

  function KeplerLayerKeyframes(_ref) {
    var _this;

    var layer = _ref.layer,
        timings = _ref.timings,
        keyframes = _ref.keyframes,
        easings = _ref.easings;
    (0, _classCallCheck2.default)(this, KeplerLayerKeyframes);
    _this = _super.call(this, {
      timings: timings,
      keyframes: keyframes,
      easings: easings,
      features: getFeatures(layer)
    });
    _this.layer = layer;
    return _this;
  }

  (0, _createClass2.default)(KeplerLayerKeyframes, [{
    key: "set",
    value: function set(_ref2) {
      var _ref2$layer = _ref2.layer,
          layer = _ref2$layer === void 0 ? undefined : _ref2$layer,
          timings = _ref2.timings,
          keyframes = _ref2.keyframes,
          easings = _ref2.easings,
          interpolators = _ref2.interpolators;
      if (layer) this.layer = layer;
      (0, _get2.default)((0, _getPrototypeOf2.default)(KeplerLayerKeyframes.prototype), "set", this).call(this, {
        timings: timings,
        keyframes: keyframes,
        easings: easings,
        interpolators: interpolators
      });
    }
  }]);
  return KeplerLayerKeyframes;
}(_keyframes.default);

var _default = KeplerLayerKeyframes;
exports.default = _default;
