// @hubble.gl/core/dist/es5/keyframes/kepler-trip-keyframes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _keyframes = _interopRequireDefault(require("./keyframes"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function tripKeyframes(_ref) {
  var animationConfig = _ref.animationConfig;
  return {
    keyframes: [{
      currentTime: animationConfig.domain[0]
    }, {
      currentTime: animationConfig.domain[1]
    }]
  };
}

var KeplerTripKeyframes = function (_Keyframes) {
  (0, _inherits2.default)(KeplerTripKeyframes, _Keyframes);

  var _super = _createSuper(KeplerTripKeyframes);

  function KeplerTripKeyframes(_ref2) {
    var _ref2$animationConfig = _ref2.animationConfig,
        animationConfig = _ref2$animationConfig === void 0 ? undefined : _ref2$animationConfig,
        timings = _ref2.timings,
        keyframes = _ref2.keyframes,
        easings = _ref2.easings,
        interpolators = _ref2.interpolators;
    (0, _classCallCheck2.default)(this, KeplerTripKeyframes);
    return _super.call(this, KeplerTripKeyframes._processParams({
      animationConfig: animationConfig,
      timings: timings,
      keyframes: keyframes,
      easings: easings,
      interpolators: interpolators
    }));
  }

  (0, _createClass2.default)(KeplerTripKeyframes, [{
    key: "set",
    value: function set(_ref3) {
      var _ref3$animationConfig = _ref3.animationConfig,
          animationConfig = _ref3$animationConfig === void 0 ? undefined : _ref3$animationConfig,
          timings = _ref3.timings,
          keyframes = _ref3.keyframes,
          easings = _ref3.easings,
          interpolators = _ref3.interpolators;
      (0, _get2.default)((0, _getPrototypeOf2.default)(KeplerTripKeyframes.prototype), "set", this).call(this, KeplerTripKeyframes._processParams({
        animationConfig: animationConfig,
        timings: timings,
        keyframes: keyframes,
        easings: easings,
        interpolators: interpolators
      }));
    }
  }], [{
    key: "_processParams",
    value: function _processParams(_ref4) {
      var _ref4$animationConfig = _ref4.animationConfig,
          animationConfig = _ref4$animationConfig === void 0 ? undefined : _ref4$animationConfig,
          timings = _ref4.timings,
          keyframes = _ref4.keyframes,
          easings = _ref4.easings,
          interpolators = _ref4.interpolators;
      var params = {
        features: ['currentTime'],
        timings: timings,
        keyframes: keyframes,
        easings: easings,
        interpolators: interpolators
      };

      if (animationConfig && keyframes === undefined) {
        if (timings.length !== 2) throw new Error('[start, end] timings required.');
        params = _objectSpread(_objectSpread({}, params), tripKeyframes({
          animationConfig: animationConfig
        }));
      }

      return params;
    }
  }]);
  return KeplerTripKeyframes;
}(_keyframes.default);

var _default = KeplerTripKeyframes;
exports.default = _default;
