// @kepler.gl/components/dist/common/time-range-slider-time-title.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _icons = require("./icons");

var _utils = require("@kepler.gl/utils");

var _templateObject;

var TimeValueWrapper = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  align-items: center;\n  font-size: ", ";\n  justify-content: ", ";\n\n  .horizontal-bar {\n    padding: 0 12px;\n    color: ", ";\n  }\n\n  .time-value {\n    display: flex;\n    flex-direction: ", ";\n    align-items: flex-start;\n    max-width: ", ";\n    span {\n      color: ", ";\n    }\n  }\n\n  .time-value:last-child {\n    align-items: flex-end;\n    text-align: right;\n  }\n"])), function (props) {
  return props.theme.timeTitleFontSize;
}, function (props) {
  return props.isEnlarged ? 'center' : 'space-between';
}, function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.isEnlarged ? 'row' : 'column';
}, function (props) {
  return !props.isEnlarged ? '40%' : 'auto';
}, function (props) {
  return props.theme.textColor;
});

var TimeValue = function TimeValue(_ref) {
  var value = _ref.value;
  return (
    /*#__PURE__*/
    // render two lines if not enlarged
    _react["default"].createElement("div", {
      className: "time-value"
    }, /*#__PURE__*/_react["default"].createElement("span", null, value))
  );
};

function TimeRangeSliderTimeTitleFactory() {
  var TimeTitle = function TimeTitle(_ref2) {
    var value = _ref2.value,
        isEnlarged = _ref2.isEnlarged,
        timezone = _ref2.timezone,
        timeFormat = _ref2.timeFormat;
    return /*#__PURE__*/_react["default"].createElement(TimeValueWrapper, {
      isEnlarged: isEnlarged,
      className: "time-range-slider__time-title"
    }, /*#__PURE__*/_react["default"].createElement(TimeValue, {
      key: 0,
      value: (0, _utils.datetimeFormatter)(timezone)(timeFormat)(value[0])
    }), isEnlarged ? /*#__PURE__*/_react["default"].createElement("div", {
      className: "horizontal-bar"
    }, /*#__PURE__*/_react["default"].createElement(_icons.Minus, {
      height: "12px"
    })) : null, /*#__PURE__*/_react["default"].createElement(TimeValue, {
      key: 1,
      value: (0, _utils.datetimeFormatter)(timezone)(timeFormat)(value[1])
    }));
  };

  return TimeTitle;
}

var _default = TimeRangeSliderTimeTitleFactory;
exports["default"] = _default;
