// @hubble.gl/react/dist/es5/components/export-video/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getResolutionSetting = getResolutionSetting;
exports.isResolution = exports.RESOLUTIONS = exports.FORMATS = exports.SIDEPANEL_WIDTH = exports.DEFAULT_FILENAME = exports.DEFAULT_SETTINGS_WIDTH = exports.DEFAULT_ROW_GAP = exports.DEFAULT_PADDING = exports.DEFAULT_BUTTON_WIDTH = exports.DEFAULT_BUTTON_HEIGHT = exports.DEFAULT_ICON_BUTTON_HEIGHT = void 0;
var DEFAULT_ICON_BUTTON_HEIGHT = '16px';
exports.DEFAULT_ICON_BUTTON_HEIGHT = DEFAULT_ICON_BUTTON_HEIGHT;
var DEFAULT_BUTTON_HEIGHT = '32px';
exports.DEFAULT_BUTTON_HEIGHT = DEFAULT_BUTTON_HEIGHT;
var DEFAULT_BUTTON_WIDTH = '64px';
exports.DEFAULT_BUTTON_WIDTH = DEFAULT_BUTTON_WIDTH;
var DEFAULT_PADDING = 32;
exports.DEFAULT_PADDING = DEFAULT_PADDING;
var DEFAULT_ROW_GAP = 16;
exports.DEFAULT_ROW_GAP = DEFAULT_ROW_GAP;
var DEFAULT_SETTINGS_WIDTH = 280;
exports.DEFAULT_SETTINGS_WIDTH = DEFAULT_SETTINGS_WIDTH;
var DEFAULT_FILENAME = 'kepler.gl';
exports.DEFAULT_FILENAME = DEFAULT_FILENAME;
var SIDEPANEL_WIDTH = 300;
exports.SIDEPANEL_WIDTH = SIDEPANEL_WIDTH;
var FORMATS = [{
  value: 'gif',
  label: 'GIF'
}, {
  value: 'webm',
  label: 'WebM Video'
}, {
  value: 'png',
  label: 'PNG Sequence'
}, {
  value: 'jpeg',
  label: 'JPEG Sequence'
}];
exports.FORMATS = FORMATS;
var RESOLUTIONS = [{
  value: '960x540',
  label: 'Good 16:9 (540p)',
  width: 960,
  height: 540
}, {
  value: '1280x720',
  label: 'High 16:9 (720p)',
  width: 1280,
  height: 720
}, {
  value: '1920x1080',
  label: 'Highest 16:9 (1080p)',
  width: 1920,
  height: 1080
}, {
  value: '640x480',
  label: 'Good 4:3 (480p)',
  width: 640,
  height: 480
}, {
  value: '1280x960',
  label: 'High 4:3 (960p)',
  width: 1280,
  height: 960
}, {
  value: '1920x1440',
  label: 'Highest 4:3 (1440p)',
  width: 1920,
  height: 1440
}];
exports.RESOLUTIONS = RESOLUTIONS;

var isResolution = function isResolution(value) {
  return function (option) {
    return option.value === value;
  };
};

exports.isResolution = isResolution;

function getResolutionSetting(value) {
  return RESOLUTIONS.find(isResolution(value)) || RESOLUTIONS[0];
}
