// @hubble.gl/react/dist/es5/components/export-video/export-video-panel-footer.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _constants = require("./constants");

var _injectKepler = require("../inject-kepler");

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  justify-content: space-between;\n  margin-top: ", "px;\n  padding: ", "px;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var PanelFooterInner = _styledComponents["default"].div(_templateObject(), _constants.DEFAULT_ROW_GAP, _constants.DEFAULT_PADDING);

var ButtonGroup = _styledComponents["default"].div(_templateObject2());

var ExportVideoPanelFooter = function ExportVideoPanelFooter(_ref) {
  var handleClose = _ref.handleClose,
      handlePreviewVideo = _ref.handlePreviewVideo,
      handleRenderVideo = _ref.handleRenderVideo;
  return _react["default"].createElement(_injectKepler.WithKeplerUI, null, function (_ref2) {
    var Button = _ref2.Button;
    return _react["default"].createElement(PanelFooterInner, {
      className: "export-video-panel__footer"
    }, _react["default"].createElement(Button, {
      width: _constants.DEFAULT_BUTTON_WIDTH,
      height: _constants.DEFAULT_BUTTON_HEIGHT,
      secondary: true,
      className: 'export-video-button',
      onClick: handlePreviewVideo
    }, "Preview"), _react["default"].createElement(ButtonGroup, null, _react["default"].createElement(Button, {
      width: _constants.DEFAULT_BUTTON_WIDTH,
      height: _constants.DEFAULT_BUTTON_HEIGHT,
      link: true,
      className: 'export-video-button',
      onClick: handleClose
    }, "Cancel"), _react["default"].createElement(Button, {
      width: _constants.DEFAULT_BUTTON_WIDTH,
      height: _constants.DEFAULT_BUTTON_HEIGHT,
      className: 'export-video-button',
      onClick: handleRenderVideo
    }, "Render")));
  });
};

var _default = (0, _styledComponents.withTheme)(ExportVideoPanelFooter);

exports["default"] = _default;
