// @hubble.gl/react/dist/es5/components/export-video/export-video-modal.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _reactModal = _interopRequireDefault(require("react-modal"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _reselect = require("reselect");

var _constants = require("./constants");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  position: relative;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var ModalContainer = _styledComponents["default"].div(_templateObject());

var ExportVideoModal = function (_Component) {
  (0, _inherits2["default"])(ExportVideoModal, _Component);

  var _super = _createSuper(ExportVideoModal);

  function ExportVideoModal() {
    var _this;

    (0, _classCallCheck2["default"])(this, ExportVideoModal);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "root", (0, _react.createRef)());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "settingsPosSelector", (0, _reselect.createSelector)(_this.showSettingsSelector, function (showSettings) {
      var theme = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var _this$props = _this.props,
          defaultSettingsPos = _this$props.defaultSettingsPos,
          bottomBuffer = _this$props.bottomBuffer,
          settingsHeight = _this$props.settingsHeight;
      if (showSettings === false || !_this.root || !_this.root.current) return defaultSettingsPos;
      var _theme$sidePanelInner = theme.sidePanelInnerPadding,
          sidePanelInnerPadding = _theme$sidePanelInner === void 0 ? 16 : _theme$sidePanelInner,
          _theme$sidePanel = theme.sidePanel,
          sidePanel = _theme$sidePanel === void 0 ? {} : _theme$sidePanel,
          _theme$sidePanelScrol = theme.sidePanelScrollBarWidth,
          sidePanelScrollBarWidth = _theme$sidePanelScrol === void 0 ? 10 : _theme$sidePanelScrol;
      var sidePanelLeft = (sidePanel.margin || {}).left || 20;
      var offsetX = sidePanelInnerPadding + sidePanelLeft + sidePanelScrollBarWidth;

      var bounding = _this.root.current.getBoundingClientRect();

      var x = bounding.x,
          y = bounding.y,
          width = bounding.width;
      var top = y + settingsHeight <= window.innerHeight - bottomBuffer ? y : window.innerHeight - bottomBuffer - settingsHeight;
      return {
        top: "".concat(top, "px"),
        left: "".concat(x + width + offsetX, "px")
      };
    }));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "modalStylesSelector", (0, _reselect.createSelector)(_this.themeSelector, _this.settingsPosSelector, function (theme, settingsPos) {
      return {
        content: {
          top: 'auto',
          left: 'auto',
          right: "calc(50% - ".concat(_constants.SIDEPANEL_WIDTH / 2, "px)"),
          bottom: '50%',
          transform: 'translate(50%, 50%)',
          padding: '0px 0px 0px 0px',
          border: 0,
          backgroundColor: theme.sidePanelBg,
          borderRadius: theme.panelBorderRadius || '2px'
        },
        overlay: {
          backgroundColor: 'rgba(0, 0, 0, .5)',
          zIndex: _this.props.theme && _this.props.theme.modalOverLayZ || 1000
        }
      };
    }));
    return _this;
  }

  (0, _createClass2["default"])(ExportVideoModal, [{
    key: "showSettingsSelector",
    value: function showSettingsSelector(props) {
      return props.showSettings;
    }
  }, {
    key: "themeSelector",
    value: function themeSelector(props) {
      return props.theme;
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          isOpen = _this$props2.isOpen,
          theme = _this$props2.theme,
          children = _this$props2.children;
      var modalStyles = this.modalStylesSelector(this.props);
      return _react["default"].createElement(_styledComponents.ThemeProvider, {
        theme: theme
      }, _react["default"].createElement(ModalContainer, {
        className: "export-video-modal",
        ref: this.root
      }, this.root.current ? _react["default"].createElement(_reactModal["default"], {
        isOpen: isOpen,
        style: modalStyles,
        ariaHideApp: false,
        parentSelector: function parentSelector() {
          return _this2.root.current || {
            removeChild: function removeChild() {},
            appendChild: function appendChild() {}
          };
        }
      }, children) : null));
    }
  }]);
  return ExportVideoModal;
}(_react.Component);

(0, _defineProperty2["default"])(ExportVideoModal, "defaultProps", {
  defaultSettingsPos: {
    top: '320px',
    left: '320px'
  },
  bottomBuffer: 212
});
var _default = ExportVideoModal;
exports["default"] = _default;
