// @hubble.gl/react/dist/es5/components/export-video/export-video-panel-settings.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _constants = require("./constants");

var _utils = require("./utils");

var _injectKepler = require("../inject-kepler");

function _templateObject5() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: grid;\n  grid-template-columns: 88px auto;\n  grid-template-rows: repeat(\n    ", ",\n    ", "\n  );\n  grid-row-gap: ", "px;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  align-self: center;\n  color: ", ";\n  font-weight: 500;\n  font-size: 11px;\n  padding: 0 12px;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  align-self: center;\n  color: ", ";\n  font-weight: 400;\n  font-size: 11px;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  position: relative;\n  flex-grow: 1;\n  margin-right: 24px;\n  margin-left: 24px;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  align-self: center;\n  color: ", ";\n  font-weight: 500;\n  font-size: 13px;\n  margin-top: ", ";\n  margin-bottom: ", "px;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var StyledSection = _styledComponents["default"].div(_templateObject(), function (props) {
  return props.theme.labelColor;
}, function (props) {
  return props.top ? '0px' : "".concat(_constants.DEFAULT_PADDING, "px");
}, _constants.DEFAULT_ROW_GAP);

var SliderWrapper = _styledComponents["default"].div(_templateObject2());

var StyledLabelCell = _styledComponents["default"].div(_templateObject3(), function (props) {
  return props.theme.labelColor;
});

var StyledValueCell = _styledComponents["default"].div(_templateObject4(), function (props) {
  return props.theme.textColor;
});

var InputGrid = _styledComponents["default"].div(_templateObject5(), function (props) {
  return props.rows;
}, function (props) {
  return props.rowHeight ? props.rowHeight : '34px';
}, _constants.DEFAULT_ROW_GAP);

var getOptionValue = function getOptionValue(r) {
  return r.value;
};

var displayOption = function displayOption(r) {
  return r.label;
};

var getSelectedItems = function getSelectedItems(options, value) {
  return options.find(function (o) {
    return o.value === value;
  });
};

var ExportVideoPanelSettings = function ExportVideoPanelSettings(_ref) {
  var setMediaType = _ref.setMediaType,
      setCameraPreset = _ref.setCameraPreset,
      setFileName = _ref.setFileName,
      setResolution = _ref.setResolution,
      settingsData = _ref.settingsData,
      durationMs = _ref.durationMs,
      frameRate = _ref.frameRate,
      resolution = _ref.resolution,
      mediaType = _ref.mediaType,
      setDuration = _ref.setDuration;
  return _react["default"].createElement(_injectKepler.WithKeplerUI, null, function (_ref2) {
    var Input = _ref2.Input,
        ItemSelector = _ref2.ItemSelector,
        Slider = _ref2.Slider;
    return _react["default"].createElement("div", null, _react["default"].createElement(StyledSection, {
      top: true
    }, "Export Settings"), _react["default"].createElement(InputGrid, {
      rows: 5
    }, _react["default"].createElement(StyledLabelCell, null, "File Name"), _react["default"].createElement(Input, {
      value: settingsData.fileName,
      placeholder: _constants.DEFAULT_FILENAME,
      onChange: function onChange(e) {
        return setFileName(e.target.value);
      }
    }), _react["default"].createElement(StyledLabelCell, null, "Media Type"), _react["default"].createElement(ItemSelector, {
      selectedItems: getSelectedItems(_constants.FORMATS, settingsData.mediaType),
      options: _constants.FORMATS,
      getOptionValue: getOptionValue,
      displayOption: displayOption,
      multiSelect: false,
      searchable: false,
      onChange: setMediaType
    }), _react["default"].createElement(StyledLabelCell, null, "Resolution"), _react["default"].createElement(ItemSelector, {
      selectedItems: getSelectedItems(_constants.RESOLUTIONS, settingsData.resolution),
      options: _constants.RESOLUTIONS,
      getOptionValue: getOptionValue,
      displayOption: displayOption,
      multiSelect: false,
      searchable: false,
      onChange: setResolution
    }), _react["default"].createElement(StyledLabelCell, null, "Duration"), _react["default"].createElement(StyledValueCell, null, _react["default"].createElement(SliderWrapper, {
      style: {
        width: '100%',
        marginLeft: '0px'
      },
      className: "modal-duration__slider"
    }, _react["default"].createElement(Slider, {
      showValues: false,
      enableBarDrag: true,
      isRanged: false,
      value1: durationMs,
      step: 100,
      minValue: 100,
      maxValue: 10000,
      style: {
        width: '70px'
      },
      onSlider1Change: function onSlider1Change(val) {
        setDuration(val);
      }
    }), _react["default"].createElement("div", {
      style: {
        alignSelf: 'center',
        paddingLeft: '8px',
        width: '56px'
      }
    }, (0, _utils.msConversion)(durationMs)))), _react["default"].createElement(StyledLabelCell, null, "File Size"), _react["default"].createElement(StyledValueCell, null, (0, _utils.estimateFileSize)(frameRate, resolution, durationMs, mediaType))), _react["default"].createElement(StyledSection, null, "Video Effects"), _react["default"].createElement(InputGrid, {
      rows: 1
    }, _react["default"].createElement(StyledLabelCell, null, "Camera"), _react["default"].createElement(ItemSelector, {
      selectedItems: settingsData.cameraPreset,
      options: ['None', 'Orbit (90ยบ)', 'Orbit (180ยบ)', 'Orbit (360ยบ)', 'North to South', 'South to North', 'East to West', 'West to East', 'Zoom Out', 'Zoom In'],
      multiSelect: false,
      searchable: false,
      onChange: setCameraPreset
    })));
  });
};

var _default = (0, _styledComponents.withTheme)(ExportVideoPanelSettings);

exports["default"] = _default;
