// @kepler.gl/components/dist/map/effects/effect-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = EffectControlFactory;

var _react = _interopRequireWildcard(require("react"));

var _icons = require("../../common/icons");

var _styledComponents = require("../../common/styled-components");

var _mapControlTooltip = _interopRequireDefault(require("../map-control-tooltip"));

// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
EffectControlFactory.deps = [_mapControlTooltip["default"]];

function EffectControlFactory(MapControlTooltip) {
  var defaultActionIcons = {
    effectsIcon: _icons.MagicWand
  };

  var EffectControl = function EffectControl(_ref) {
    var _mapControls$effect, _mapControls$effect2;

    var mapControls = _ref.mapControls,
        onToggleMapControl = _ref.onToggleMapControl,
        _ref$actionIcons = _ref.actionIcons,
        actionIcons = _ref$actionIcons === void 0 ? defaultActionIcons : _ref$actionIcons;
    var onClick = (0, _react.useCallback)(function (event) {
      event.preventDefault();
      onToggleMapControl('effect');
    }, [onToggleMapControl]);
    var showControl = mapControls === null || mapControls === void 0 ? void 0 : (_mapControls$effect = mapControls.effect) === null || _mapControls$effect === void 0 ? void 0 : _mapControls$effect.show;

    if (!showControl) {
      return null;
    }

    var active = mapControls === null || mapControls === void 0 ? void 0 : (_mapControls$effect2 = mapControls.effect) === null || _mapControls$effect2 === void 0 ? void 0 : _mapControls$effect2.active;
    return /*#__PURE__*/_react["default"].createElement(MapControlTooltip, {
      id: "show-effect",
      message: active ? 'tooltip.hideEffectPanel' : 'tooltip.showEffectPanel'
    }, /*#__PURE__*/_react["default"].createElement(_styledComponents.MapControlButton, {
      className: "map-control-button toggle-effect",
      onClick: onClick,
      active: active
    }, /*#__PURE__*/_react["default"].createElement(actionIcons.effectsIcon, {
      height: "22px"
    })));
  };

  EffectControl.displayName = 'EffectControl';
  return /*#__PURE__*/_react["default"].memo(EffectControl);
}
