// kepler.gl/src/constants/dist/dnd-layer-items.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SORTABLE_EFFECT_TYPE = exports.SORTABLE_EFFECT_PANEL_TYPE = exports.SORTABLE_LAYER_TYPE = exports.DROPPABLE_MAP_CONTAINER_TYPE = exports.SORTABLE_SIDE_PANEL_TYPE = exports.DND_EMPTY_MODIFIERS = exports.DND_MODIFIERS = void 0;

var _modifiers = require("@dnd-kit/modifiers");

// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var DND_MODIFIERS = [_modifiers.restrictToVerticalAxis];
exports.DND_MODIFIERS = DND_MODIFIERS;
var DND_EMPTY_MODIFIERS = [];
exports.DND_EMPTY_MODIFIERS = DND_EMPTY_MODIFIERS;
var SORTABLE_SIDE_PANEL_TYPE = 'root';
exports.SORTABLE_SIDE_PANEL_TYPE = SORTABLE_SIDE_PANEL_TYPE;
var DROPPABLE_MAP_CONTAINER_TYPE = 'map';
exports.DROPPABLE_MAP_CONTAINER_TYPE = DROPPABLE_MAP_CONTAINER_TYPE;
var SORTABLE_LAYER_TYPE = 'layer';
exports.SORTABLE_LAYER_TYPE = SORTABLE_LAYER_TYPE;
var SORTABLE_EFFECT_PANEL_TYPE = 'root';
exports.SORTABLE_EFFECT_PANEL_TYPE = SORTABLE_EFFECT_PANEL_TYPE;
var SORTABLE_EFFECT_TYPE = 'effect';
exports.SORTABLE_EFFECT_TYPE = SORTABLE_EFFECT_TYPE;
