// @kepler.gl/utils/dist/h3-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getVertices = getVertices;
exports.getCentroid = getCentroid;
exports.idToPolygonGeo = idToPolygonGeo;
Object.defineProperty(exports, "h3GetResolution", {
  enumerable: true,
  get: function get() {
    return _h3Js.h3GetResolution;
  }
});
Object.defineProperty(exports, "h3IsValid", {
  enumerable: true,
  get: function get() {
    return _h3Js.h3IsValid;
  }
});
exports.getHexFields = exports.isHexField = void 0;

var _h3Js = require("h3-js");

var _constants = require("@kepler.gl/constants");

var _dataUtils = require("./data-utils");

// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
// get vertices should return [lon, lat]
function getVertices(_ref) {
  var id = _ref.id;
  // always reverse it
  return (0, _h3Js.h3ToGeoBoundary)(id, true);
} // get centroid should return [lon, lat]


function getCentroid(_ref2) {
  var id = _ref2.id;
  // always reverse it to [lng, lat]
  return (0, _h3Js.h3ToGeo)(id).reverse();
}

function idToPolygonGeo(object, properties) {
  if (!(object !== null && object !== void 0 && object.id)) {
    return null;
  }

  var vertices = getVertices(object);
  return {
    type: 'Feature',
    geometry: {
      coordinates: properties !== null && properties !== void 0 && properties.isClosed ? [vertices] : vertices,
      type: properties !== null && properties !== void 0 && properties.isClosed ? 'Polygon' : 'LineString'
    },
    properties: properties
  };
}

var isHexField = function isHexField(field, fieldIdx, dataContainer) {
  if (field.type !== _constants.ALL_FIELD_TYPES.string) {
    return false;
  }

  var firstDP = dataContainer.find(function (d) {
    return (0, _dataUtils.notNullorUndefined)(d.valueAt(fieldIdx));
  }, true);
  return firstDP && (0, _h3Js.h3IsValid)(firstDP.valueAt(fieldIdx));
};

exports.isHexField = isHexField;

var getHexFields = function getHexFields(fields, dataContainer) {
  return fields.filter(function (f, i) {
    return isHexField(f, i, dataContainer);
  });
};

exports.getHexFields = getHexFields;
