// @kepler.gl/components/dist/map/map-popover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getSelectedFeature = getSelectedFeature;
exports["default"] = MapPopoverFactory;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _mapPopoverContent = _interopRequireDefault(require("./map-popover-content"));

var _icons = require("../common/icons");

var _reactIntl = require("react-intl");

var _localization = require("@kepler.gl/localization");

var _headless = _interopRequireDefault(require("@tippyjs/react/headless"));

var _ = require("../");

var _layers = require("@kepler.gl/layers");

var _utils = require("@kepler.gl/utils");

var _constants = require("@kepler.gl/constants");

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var SELECTABLE_LAYERS = [_constants.LAYER_TYPES.hexagonId, _constants.LAYER_TYPES.geojson];
var MAX_WIDTH = 500;
var MAX_HEIGHT = 600;

var StyledMapPopover = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: column;\n  max-width: ", "px;\n  max-height: ", "px;\n  padding: 14px;\n  & > * + * {\n    margin-top: 6px;\n  }\n  ", ";\n  font-family: ", ";\n  font-size: 11px;\n  font-weight: 500;\n  background-color: ", ";\n  color: ", ";\n  z-index: 1000;\n  overflow-x: auto;\n  box-shadow: ", ";\n\n  :hover {\n    background-color: ", ";\n  }\n\n  .primary-label {\n    color: ", ";\n    font-size: 10px;\n  }\n\n  .map-popover__layer-info,\n  .coordingate-hover-info {\n    & > * + * {\n      margin-top: 7px;\n    }\n  }\n\n  table {\n    width: auto;\n    display: grid;\n    border-collapse: collapse;\n    row-gap: 5px;\n    column-gap: 5px;\n  }\n\n  .coordingate-hover-info > table {\n    grid-template-columns: auto auto auto;\n  }\n  .map-popover__layer-info > table {\n    grid-template-columns: auto auto;\n  }\n\n  tbody,\n  tr {\n    display: contents;\n  }\n\n  td {\n    border-color: transparent;\n    color: ", ";\n  }\n\n  td.row__value {\n    text-align: right;\n    font-weight: 500;\n    color: ", ";\n  }\n"])), MAX_WIDTH, MAX_HEIGHT, function (props) {
  return props.theme.scrollBar;
}, function (props) {
  return props.theme.fontFamily;
}, function (props) {
  return props.theme.panelBackground;
}, function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.theme.panelBoxShadow;
}, function (props) {
  return "".concat(props.theme.panelBackground, "dd");
}, function (props) {
  return props.theme.notificationColors.success;
}, function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.theme.textColorHl;
});

var PinnedButtons = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  align-self: center;\n  align-items: center;\n  justify-items: center;\n  & > * + * {\n    margin-left: 10px;\n  }\n"])));

var PopoverContent = _styledComponents["default"].div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: column;\n  & > * + * {\n    margin-top: 12px;\n  }\n"])));

var StyledIcon = _styledComponents["default"].div(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n\n  :hover {\n    cursor: pointer;\n    color: ", ";\n  }\n"])), function (props) {
  return props.theme.activeColor;
}, function (props) {
  return props.theme.linkBtnColor;
});

var StyledSelectGeometry = _styledComponents["default"].div(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  align-items: center;\n  color: ", ";\n  svg {\n    margin-right: 6px;\n  }\n\n  :hover {\n    cursor: pointer;\n    color: ", ";\n  }\n"])), function (props) {
  return props.theme.textColorHl;
}, function (props) {
  return props.theme.linkBtnColor;
});

MapPopoverFactory.deps = [_mapPopoverContent["default"]];

function createVirtualReference(container, x, y) {
  var size = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;
  var bounds = container && container.getBoundingClientRect ? container.getBoundingClientRect() : {};
  var left = (bounds.left || 0) + x - size / 2;
  var top = (bounds.top || 0) + y - size / 2;
  return {
    left: left,
    top: top,
    right: left + size,
    bottom: top + size,
    width: size,
    height: size,
    // These properties are present to meet the DOMRect interface
    y: top,
    x: left,
    toJSON: function toJSON() {
      return this;
    }
  };
}

function getOffsetForPlacement(_ref) {
  var placement = _ref.placement,
      reference = _ref.reference,
      popper = _ref.popper;
  var gap = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 20;

  switch (placement) {
    case 'top-start':
    case 'bottom-start':
      return [gap, gap];

    case 'top-end':
    case 'bottom-end':
      return [-gap, gap];

    default:
      return [0, 0];
  }
}

function getPopperOptions(container) {
  return {
    modifiers: [{
      name: 'preventOverflow',
      options: {
        boundary: container
      }
    }]
  };
}

function getSelectedFeature(layerHoverProp) {
  var _layer$config, _layer$config$columns, _layer$config$columns2, _layerHoverProp$data, _layer$config2, _layer$config2$column, _layer$config2$column2, _layerHoverProp$data2;

  var layer = layerHoverProp === null || layerHoverProp === void 0 ? void 0 : layerHoverProp.layer;
  var fieldIdx;
  var selectedFeature;

  switch (layer === null || layer === void 0 ? void 0 : layer.type) {
    case _constants.LAYER_TYPES.hexagonId:
      fieldIdx = (_layer$config = layer.config) === null || _layer$config === void 0 ? void 0 : (_layer$config$columns = _layer$config.columns) === null || _layer$config$columns === void 0 ? void 0 : (_layer$config$columns2 = _layer$config$columns.hex_id) === null || _layer$config$columns2 === void 0 ? void 0 : _layer$config$columns2.fieldIdx;
      selectedFeature = (0, _utils.idToPolygonGeo)({
        id: layerHoverProp === null || layerHoverProp === void 0 ? void 0 : (_layerHoverProp$data = layerHoverProp.data) === null || _layerHoverProp$data === void 0 ? void 0 : _layerHoverProp$data[fieldIdx]
      }, {
        isClosed: true
      });
      break;

    case _constants.LAYER_TYPES.geojson:
      fieldIdx = (_layer$config2 = layer.config) === null || _layer$config2 === void 0 ? void 0 : (_layer$config2$column = _layer$config2.columns) === null || _layer$config2$column === void 0 ? void 0 : (_layer$config2$column2 = _layer$config2$column.geojson) === null || _layer$config2$column2 === void 0 ? void 0 : _layer$config2$column2.fieldIdx;
      selectedFeature = (0, _layers.parseGeoJsonRawFeature)(layerHoverProp === null || layerHoverProp === void 0 ? void 0 : (_layerHoverProp$data2 = layerHoverProp.data) === null || _layerHoverProp$data2 === void 0 ? void 0 : _layerHoverProp$data2[fieldIdx]);
      break;

    default:
      break;
  }

  return _objectSpread(_objectSpread({}, selectedFeature), {}, {
    // unique id should be assigned to features in the editor
    id: (0, _utils.generateHashId)(8)
  });
}

function MapPopoverFactory(MapPopoverContent) {
  var MapPopover = function MapPopover(_ref2) {
    var x = _ref2.x,
        y = _ref2.y,
        frozen = _ref2.frozen,
        coordinate = _ref2.coordinate,
        layerHoverProp = _ref2.layerHoverProp,
        isBase = _ref2.isBase,
        zoom = _ref2.zoom,
        container = _ref2.container,
        onClose = _ref2.onClose,
        onSetFeatures = _ref2.onSetFeatures,
        setSelectedFeature = _ref2.setSelectedFeature,
        featureCollection = _ref2.featureCollection;

    var _useState = (0, _react.useState)('start'),
        _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
        horizontalPlacement = _useState2[0],
        setHorizontalPlacement = _useState2[1];

    var moveLeft = function moveLeft() {
      return setHorizontalPlacement('end');
    };

    var moveRight = function moveRight() {
      return setHorizontalPlacement('start');
    };

    var onSetSelectedFeature = (0, _react.useCallback)(function () {
      var clickContext = {
        mapIndex: 0,
        rightClick: true,
        position: {
          x: x,
          y: y
        }
      };
      var selectedFeature = getSelectedFeature(layerHoverProp);

      if (selectedFeature) {
        setSelectedFeature(selectedFeature, clickContext);
        var updatedFeatures = featureCollection ? [].concat((0, _toConsumableArray2["default"])(featureCollection.features), [selectedFeature]) : [selectedFeature];
        onSetFeatures(updatedFeatures);
      }

      onClose();
    }, [onClose, onSetFeatures, x, y, setSelectedFeature, layerHoverProp, featureCollection]);
    return /*#__PURE__*/_react["default"].createElement(_.RootContext.Consumer, null, function (context) {
      return /*#__PURE__*/_react["default"].createElement(_headless["default"], {
        popperOptions: getPopperOptions(container),
        zIndex: 98
        /* should be below side panel's z-index of 99  */
        ,
        visible: true,
        interactive: true // @ts-ignore
        ,
        getReferenceClientRect: function getReferenceClientRect() {
          return createVirtualReference(container, x, y);
        } // @ts-ignore
        ,
        placement: "bottom-".concat(horizontalPlacement) // @ts-ignore
        ,
        offset: getOffsetForPlacement,
        appendTo: (context === null || context === void 0 ? void 0 : context.current) || document.body,
        render: function render(attrs) {
          var _layerHoverProp$layer, _layerHoverProp$layer2;

          return /*#__PURE__*/_react["default"].createElement(StyledMapPopover, (0, _extends2["default"])({}, attrs, {
            className: "map-popover"
          }), frozen ? /*#__PURE__*/_react["default"].createElement(PinnedButtons, null, horizontalPlacement === 'start' && /*#__PURE__*/_react["default"].createElement(StyledIcon, {
            className: "popover-arrow-left",
            onClick: moveLeft
          }, /*#__PURE__*/_react["default"].createElement(_icons.ArrowLeft, null)), /*#__PURE__*/_react["default"].createElement(StyledIcon, {
            className: "popover-pin",
            onClick: onClose
          }, /*#__PURE__*/_react["default"].createElement(_icons.Pin, {
            height: "16px"
          })), horizontalPlacement === 'end' && /*#__PURE__*/_react["default"].createElement(StyledIcon, {
            className: "popover-arrow-right",
            onClick: moveRight
          }, /*#__PURE__*/_react["default"].createElement(_icons.ArrowRight, null)), isBase && /*#__PURE__*/_react["default"].createElement("div", {
            className: "primary-label"
          }, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
            id: "mapPopover.primary"
          }))) : null, /*#__PURE__*/_react["default"].createElement(PopoverContent, null, /*#__PURE__*/_react["default"].createElement(MapPopoverContent, {
            coordinate: coordinate,
            zoom: zoom,
            layerHoverProp: layerHoverProp
          })), layerHoverProp !== null && layerHoverProp !== void 0 && (_layerHoverProp$layer = layerHoverProp.layer) !== null && _layerHoverProp$layer !== void 0 && _layerHoverProp$layer.type && SELECTABLE_LAYERS.includes(layerHoverProp === null || layerHoverProp === void 0 ? void 0 : (_layerHoverProp$layer2 = layerHoverProp.layer) === null || _layerHoverProp$layer2 === void 0 ? void 0 : _layerHoverProp$layer2.type) && frozen ? /*#__PURE__*/_react["default"].createElement(StyledSelectGeometry, {
            className: "select-geometry",
            onClick: onSetSelectedFeature
          }, /*#__PURE__*/_react["default"].createElement(_icons.CursorPoint, null), "Select Geometry") : null);
        }
      });
    });
  };

  return (0, _reactIntl.injectIntl)(MapPopover);
}
