// @hubble.gl/react/dist/es6/components/export-video/export-video-panel-footer.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/esm/taggedTemplateLiteral";

function _templateObject2() {
  const data = _taggedTemplateLiteral(["\n  display: flex;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  const data = _taggedTemplateLiteral(["\n  display: flex;\n  justify-content: space-between;\n  margin-top: ", "px;\n  padding: ", "px;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

import React from 'react';
import styled, { withTheme } from 'styled-components';
import { DEFAULT_PADDING, DEFAULT_ROW_GAP, DEFAULT_BUTTON_HEIGHT, DEFAULT_BUTTON_WIDTH } from './constants';
import { WithKeplerUI } from '../inject-kepler';
const PanelFooterInner = styled.div(_templateObject(), DEFAULT_ROW_GAP, DEFAULT_PADDING);
const ButtonGroup = styled.div(_templateObject2());

const ExportVideoPanelFooter = ({
  handleClose,
  handlePreviewVideo,
  handleRenderVideo
}) => React.createElement(WithKeplerUI, null, ({
  Button
}) => React.createElement(PanelFooterInner, {
  className: "export-video-panel__footer"
}, React.createElement(Button, {
  width: DEFAULT_BUTTON_WIDTH,
  height: DEFAULT_BUTTON_HEIGHT,
  secondary: true,
  className: 'export-video-button',
  onClick: handlePreviewVideo
}, "Preview"), React.createElement(ButtonGroup, null, React.createElement(Button, {
  width: DEFAULT_BUTTON_WIDTH,
  height: DEFAULT_BUTTON_HEIGHT,
  link: true,
  className: 'export-video-button',
  onClick: handleClose
}, "Cancel"), React.createElement(Button, {
  width: DEFAULT_BUTTON_WIDTH,
  height: DEFAULT_BUTTON_HEIGHT,
  className: 'export-video-button',
  onClick: handleRenderVideo
}, "Render"))));

export default withTheme(ExportVideoPanelFooter);
