// @hubble.gl/core/dist/esm/encoders/video/webm-encoder.js
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import WebMWriter from 'webm-writer';
import FrameEncoder from '../frame-encoder';

var WebMEncoder = function (_FrameEncoder) {
  _inherits(WebMEncoder, _FrameEncoder);

  var _super = _createSuper(WebMEncoder);

  function WebMEncoder(settings) {
    var _this;

    _classCallCheck(this, WebMEncoder);

    _this = _super.call(this, settings);

    _defineProperty(_assertThisInitialized(_this), "videoWriter", void 0);

    _this.quality = 0.8;

    if (settings.webm && settings.webm.quality) {
      _this.quality = settings.webm.quality;
    }

    var canvas = document.createElement('canvas');

    if (canvas.toDataURL('image/webp').substr(5, 10) !== 'image/webp') {
      console.error('WebP not supported - try another export format');
    }

    _this.extension = '.webm';
    _this.mimeType = 'video/webm';
    _this.videoWriter = null;
    _this.start = _this.start.bind(_assertThisInitialized(_this));
    _this.add = _this.add.bind(_assertThisInitialized(_this));
    _this.save = _this.save.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(WebMEncoder, [{
    key: "start",
    value: function start() {
      this.videoWriter = new WebMWriter({
        quality: this.quality,
        fileWriter: null,
        fd: null,
        frameRate: this.framerate
      });
    }
  }, {
    key: "add",
    value: function () {
      var _add = _asyncToGenerator(_regeneratorRuntime.mark(function _callee(canvas) {
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                this.videoWriter.addFrame(canvas);
                _context.next = 3;
                return Promise.resolve();

              case 3:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function add(_x) {
        return _add.apply(this, arguments);
      }

      return add;
    }()
  }, {
    key: "save",
    value: function () {
      var _save = _asyncToGenerator(_regeneratorRuntime.mark(function _callee2() {
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                return _context2.abrupt("return", this.videoWriter.complete());

              case 1:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function save() {
        return _save.apply(this, arguments);
      }

      return save;
    }()
  }]);

  return WebMEncoder;
}(FrameEncoder);

export default WebMEncoder;
