// @hubble.gl/react/dist/es5/components/resolution-guide.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ResolutionGuide;

var _react = _interopRequireDefault(require("react"));

function ResolutionGuide() {
  return _react["default"].createElement("div", {
    style: {
      position: 'relative'
    }
  }, _react["default"].createElement(Outline, {
    width: 2560,
    height: 1440,
    name: '2160p, 2K'
  }), _react["default"].createElement(Outline, {
    width: 2560,
    height: 1440,
    name: '1440p'
  }), _react["default"].createElement(Outline, {
    width: 1920,
    height: 1080,
    name: '1080p, Full HD'
  }), _react["default"].createElement(Outline, {
    width: 1280,
    height: 720,
    name: '720p, HD'
  }), _react["default"].createElement(Outline, {
    width: 854,
    height: 480,
    name: '480p, WSD'
  }), _react["default"].createElement(Outline, {
    width: 640,
    height: 480,
    name: '480p, SD'
  }), _react["default"].createElement(Outline, {
    width: 640,
    height: 360,
    name: '360p'
  }), _react["default"].createElement(Outline, {
    width: 426,
    height: 240,
    name: '240p'
  }));
}

function Outline(_ref) {
  var width = _ref.width,
      height = _ref.height,
      name = _ref.name;
  return _react["default"].createElement("div", {
    style: {
      position: 'absolute',
      width: width,
      height: height,
      outline: '2px dashed black'
    }
  }, _react["default"].createElement("div", {
    style: {
      position: 'absolute',
      bottom: -20,
      right: 0,
      fontSize: '14px',
      font: "normal 14px/20px 'Uber Move',Helvetica,Arial,sans-serif"
    }
  }, name, " (", width, " x ", height, ")"));
}
