// @kepler.gl/effects/dist/lighting-effect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _core = require("@deck.gl/core");

var _momentTimezone = _interopRequireDefault(require("moment-timezone"));

var _constants = require("@kepler.gl/constants");

var _utils = require("@kepler.gl/utils");

var _effect = _interopRequireDefault(require("./effect"));

var _customDeckLightingEffect = _interopRequireDefault(require("./custom-deck-lighting-effect"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var LIGHT_AND_SHADOW_EFFECT_DESC = _objectSpread(_objectSpread({}, _constants.LIGHT_AND_SHADOW_EFFECT), {}, {
  "class": null
});

var LightingEffect = /*#__PURE__*/function (_Effect) {
  (0, _inherits2["default"])(LightingEffect, _Effect);

  var _super = _createSuper(LightingEffect);

  // deckEffect: PostProcessEffect | LightingEffect | null;
  function LightingEffect(props) {
    (0, _classCallCheck2["default"])(this, LightingEffect);
    return _super.call(this, props);
  }

  (0, _createClass2["default"])(LightingEffect, [{
    key: "_initializeEffect",
    value: function _initializeEffect() {
      this.parameters = _objectSpread(_objectSpread({}, _constants.DEFAULT_LIGHT_AND_SHADOW_PROPS), {}, {
        timezone: _momentTimezone["default"].tz.guess(true)
      }, this.parameters);
      var parameters = this.parameters;
      var ambientLight = new _core.AmbientLight({
        color: parameters.ambientLightColor,
        intensity: parameters.ambientLightIntensity
      });
      var sunLight = new _core._SunLight({
        timestamp: parameters.timestamp,
        color: parameters.sunLightColor,
        intensity: parameters.sunLightIntensity,
        _shadow: true
      });
      this.deckEffect = new _customDeckLightingEffect["default"]({
        ambientLight: ambientLight,
        sunLight: sunLight
      });

      if (this.deckEffect) {
        this.deckEffect.shadowColor = [].concat((0, _toConsumableArray2["default"])((0, _utils.normalizeColor)(parameters.shadowColor)), [parameters.shadowIntensity]);
      }
    }
  }, {
    key: "getDefaultProps",
    value: function getDefaultProps() {
      var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      return (0, _get2["default"])((0, _getPrototypeOf2["default"])(LightingEffect.prototype), "getDefaultProps", this).call(this, _objectSpread({
        type: LIGHT_AND_SHADOW_EFFECT_DESC.type
      }, props));
    }
  }, {
    key: "setProps",
    value: function setProps(props) {
      (0, _get2["default"])((0, _getPrototypeOf2["default"])(LightingEffect.prototype), "setProps", this).call(this, props); // any uniform updated?

      if (props.parameters) {
        var parameters = this.parameters;

        if (this.type === LIGHT_AND_SHADOW_EFFECT_DESC.type) {
          /** @type {LightingEffect} */
          var effect = this.deckEffect;

          if (effect) {
            effect.shadowColor = [].concat((0, _toConsumableArray2["default"])((0, _utils.normalizeColor)(parameters.shadowColor)), [parameters.shadowIntensity]);
            effect.ambientLight.intensity = parameters.ambientLightIntensity;
            effect.ambientLight.color = parameters.ambientLightColor.slice();
            var sunLight = effect.directionalLights[0];

            if (sunLight) {
              sunLight.intensity = parameters.sunLightIntensity;
              sunLight.color = parameters.sunLightColor.slice();
              sunLight.timestamp = parameters.timestamp;
            }
          }
        }
      }
    }
  }]);
  return LightingEffect;
}(_effect["default"]);

var _default = LightingEffect;
exports["default"] = _default;
