// @kepler.gl/components/dist/side-panel/layer-panel/dimension-scale-selector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = require("../../common/styled-components");

var _itemSelector = _interopRequireDefault(require("../../common/item-selector/item-selector"));

var _localization = require("@kepler.gl/localization");

var _utils = require("@kepler.gl/utils");

// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var DimensionScaleSelector = function DimensionScaleSelector(_ref) {
  var label = _ref.label,
      onSelect = _ref.onSelect,
      options = _ref.options,
      scaleType = _ref.scaleType,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled;
  return /*#__PURE__*/_react["default"].createElement(_styledComponents.SidePanelSection, null, /*#__PURE__*/_react["default"].createElement(_styledComponents.PanelLabel, null, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
    id: label ? "scale.".concat((0, _utils.camelize)(label)) : 'misc.scale',
    defaultMessage: label
  })), /*#__PURE__*/_react["default"].createElement(_itemSelector["default"], {
    disabled: disabled,
    selectedItems: scaleType,
    options: options,
    multiSelect: false,
    searchable: false,
    onChange: onSelect
  }));
};

var _default = DimensionScaleSelector;
exports["default"] = _default;
