// kepler.gl/src/components/dist/common/animation-control/window-action-control.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _localization = require("@kepler.gl/localization");

var _styledComponents = require("../styled-components");

var _iconButton = _interopRequireDefault(require("../icon-button"));

// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var DELAY_SHOW = 500;

function WindowActionControlFactory() {
  var WindowActionComponent = function WindowActionComponent(_ref) {
    var toggleAnimationWindowControl = _ref.toggleAnimationWindowControl,
        showAnimationWindowControl = _ref.showAnimationWindowControl,
        btnStyle = _ref.btnStyle,
        animationItems = _ref.animationItems,
        animationWindow = _ref.animationWindow,
        buttonHeight = _ref.buttonHeight,
        setFilterAnimationWindow = _ref.setFilterAnimationWindow;
    var icon = (0, _react.useMemo)(function () {
      if (animationItems[animationWindow]) {
        var WindowIcon = animationItems[animationWindow].icon;
        return /*#__PURE__*/_react["default"].createElement(WindowIcon, {
          height: buttonHeight
        });
      }

      return null;
    }, [animationItems, animationWindow, buttonHeight]);
    return setFilterAnimationWindow ? /*#__PURE__*/_react["default"].createElement(_iconButton["default"], (0, _extends2["default"])({
      "data-tip": true,
      "data-for": "animate-window",
      className: (0, _classnames["default"])('playback-control-button', {
        active: showAnimationWindowControl
      }),
      onClick: toggleAnimationWindowControl
    }, btnStyle), icon, animationItems[animationWindow] && animationItems[animationWindow].tooltip ? /*#__PURE__*/_react["default"].createElement(_styledComponents.Tooltip, {
      id: "animate-window",
      place: "top",
      delayShow: DELAY_SHOW,
      effect: "solid"
    }, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: animationItems[animationWindow].tooltip
    })) : null) : null;
  };

  return WindowActionComponent;
}

var _default = WindowActionControlFactory;
exports["default"] = _default;
