// kepler.gl/src/components/dist/side-panel/cloud-storage-dropdown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _ = require("../");

// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var CloudStorageItems = [{
  label: 'Save',
  icon: _.Icons.Save2,
  key: 'data',
  onClick: function onClick(props) {
    return props.onSaveMap;
  }
}, {
  label: 'Settings',
  icon: _.Icons.Gear,
  key: 'settings',
  onClick: function onClick(props) {
    return props.onExportData;
  }
}];
var PanelHeaderDropdown = (0, _.PanelHeaderDropdownFactory)();

var CloudStorageDropdown = function CloudStorageDropdown(_ref) {
  var show = _ref.show,
      onClose = _ref.onClose;
  return /*#__PURE__*/_react["default"].createElement(PanelHeaderDropdown, {
    items: CloudStorageItems,
    show: show,
    onClose: onClose,
    id: "cloud-storage"
  });
};

var _default = CloudStorageDropdown;
exports["default"] = _default;
