// @hubble.gl/react/dist/esm/components/export-video/export-video-panel-preview.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import DeckGL from '@deck.gl/react';
import { StaticMap } from 'react-map-gl';
import { MapboxLayer } from '@deck.gl/mapbox';
export var ExportVideoPanelPreview = function (_Component) {
  _inherits(ExportVideoPanelPreview, _Component);

  var _super = _createSuper(ExportVideoPanelPreview);

  function ExportVideoPanelPreview(props) {
    var _this;

    _classCallCheck(this, ExportVideoPanelPreview);

    _this = _super.call(this, props);
    var user = _this.props.mapData.mapStyle.bottomMapStyle.owner;
    var mapId = _this.props.mapData.mapStyle.bottomMapStyle.id;
    _this.mapRef = React.createRef();
    _this.deckRef = React.createRef();
    _this.state = {
      timestamp: {
        latitude: 47.65,
        longitude: 7
      },
      mapStyle: "mapbox://styles/".concat(user, "/").concat(mapId),
      glContext: undefined,
      memoDevicePixelRatio: window.devicePixelRatio
    };
    _this._onLayerSetDomain = _this._onLayerSetDomain.bind(_assertThisInitialized(_this));
    _this._renderLayer = _this._renderLayer.bind(_assertThisInitialized(_this));
    _this._onMapLoad = _this._onMapLoad.bind(_assertThisInitialized(_this));
    _this._resizeVideo = _this._resizeVideo.bind(_assertThisInitialized(_this));
    _this._getContainerHeight = _this._getContainerHeight.bind(_assertThisInitialized(_this));

    _this._resizeVideo();

    return _this;
  }

  _createClass(ExportVideoPanelPreview, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.resolution !== this.props.resolution) {
        this._resizeVideo();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var memoDevicePixelRatio = this.state.memoDevicePixelRatio;

      this._setDevicePixelRatio(memoDevicePixelRatio);
    }
  }, {
    key: "_resizeVideo",
    value: function _resizeVideo() {
      var _this$props = this.props,
          exportVideoWidth = _this$props.exportVideoWidth,
          resolution = _this$props.resolution;

      this._setDevicePixelRatio(resolution[0] / exportVideoWidth);

      if (this.mapRef.current) {
        var map = this.mapRef.current.getMap();
        map.resize();
      }
    }
  }, {
    key: "_setDevicePixelRatio",
    value: function _setDevicePixelRatio(devicePixelRatio) {
      window.devicePixelRatio = devicePixelRatio;
    }
  }, {
    key: "_onLayerSetDomain",
    value: function _onLayerSetDomain(idx, colorDomain) {}
  }, {
    key: "_renderLayer",
    value: function _renderLayer(overlays, idx) {
      var _this2 = this;

      var _this$props$mapData = this.props.mapData,
          visState = _this$props$mapData.visState,
          mapState = _this$props$mapData.mapState;
      var datasets = visState.datasets,
          layers = visState.layers,
          layerData = visState.layerData,
          hoverInfo = visState.hoverInfo,
          clicked = visState.clicked,
          interactionConfig = visState.interactionConfig,
          animationConfig = visState.animationConfig;
      var layer = layers[idx];
      var data = layerData[idx];

      var _ref = datasets[layer.config.dataId] || {},
          gpuFilter = _ref.gpuFilter;

      var objectHovered = clicked || hoverInfo;
      var layerCallbacks = {
        onSetLayerDomain: function onSetLayerDomain(val) {
          return _this2._onLayerSetDomain(idx, val);
        }
      };
      var layerOverlay = layer.renderLayer({
        data: data,
        gpuFilter: gpuFilter,
        idx: idx,
        interactionConfig: interactionConfig,
        layerCallbacks: layerCallbacks,
        mapState: mapState,
        animationConfig: animationConfig,
        objectHovered: objectHovered
      });
      return overlays.concat(layerOverlay || []);
    }
  }, {
    key: "_getContainerHeight",
    value: function _getContainerHeight() {
      var _this$props2 = this.props,
          exportVideoWidth = _this$props2.exportVideoWidth,
          resolution = _this$props2.resolution;
      var aspectRatio = resolution[0] / resolution[1];
      return exportVideoWidth / aspectRatio;
    }
  }, {
    key: "_onMapLoad",
    value: function _onMapLoad() {
      var _this3 = this;

      var map = this.mapRef.current.getMap();
      var deck = this.deckRef.current.deck;
      map.addLayer(new MapboxLayer({
        id: 'my-deck',
        deck: deck
      }));
      map.on('render', function () {
        return _this3.props.adapter.onAfterRender(function () {
          _this3.forceUpdate();
        });
      });
      map.resize();
    }
  }, {
    key: "render",
    value: function render() {
      var _this4 = this;

      var layerOrder = this.props.mapData.visState.layerOrder;
      var deckGlLayers = [];

      if (layerOrder && layerOrder.length) {
        deckGlLayers = layerOrder.slice().reverse().reduce(this._renderLayer, []);
      }

      var deckStyle = {
        width: '100%',
        height: '100%'
      };
      var containerStyle = {
        width: "".concat(this.props.exportVideoWidth, "px"),
        height: "".concat(this._getContainerHeight(), "px"),
        position: 'relative',
        overflow: 'auto'
      };
      return React.createElement("div", {
        id: "deck-canvas",
        style: containerStyle
      }, React.createElement(DeckGL, _extends({
        ref: this.deckRef,
        viewState: this.props.viewState,
        id: "hubblegl-overlay",
        layers: deckGlLayers,
        style: deckStyle,
        controller: true,
        glOptions: {
          stencil: true
        },
        onWebGLInitialized: function onWebGLInitialized(gl) {
          return _this4.setState({
            glContext: gl
          });
        },
        onViewStateChange: this.props.setViewState
      }, this.props.adapter.getProps(this.deckRef, function () {})), this.state.glContext && React.createElement(StaticMap, {
        ref: this.mapRef,
        mapStyle: this.state.mapStyle,
        preventStyleDiffing: true,
        gl: this.state.glContext,
        onLoad: this._onMapLoad
      })));
    }
  }]);

  return ExportVideoPanelPreview;
}(Component);
