// @hubble.gl/react/dist/esm/components/export-video/export-video-panel-footer.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/esm/taggedTemplateLiteral";

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n  justify-content: space-between;\n  margin-top: ", "px;\n  padding: ", "px;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import React from 'react';
import styled, { withTheme } from 'styled-components';
import { DEFAULT_PADDING, DEFAULT_ROW_GAP, DEFAULT_BUTTON_HEIGHT, DEFAULT_BUTTON_WIDTH } from './constants';
import { WithKeplerUI } from '../inject-kepler';
var PanelFooterInner = styled.div(_templateObject(), DEFAULT_ROW_GAP, DEFAULT_PADDING);
var ButtonGroup = styled.div(_templateObject2());

var ExportVideoPanelFooter = function ExportVideoPanelFooter(_ref) {
  var handleClose = _ref.handleClose,
      handlePreviewVideo = _ref.handlePreviewVideo,
      handleRenderVideo = _ref.handleRenderVideo;
  return React.createElement(WithKeplerUI, null, function (_ref2) {
    var Button = _ref2.Button;
    return React.createElement(PanelFooterInner, {
      className: "export-video-panel__footer"
    }, React.createElement(Button, {
      width: DEFAULT_BUTTON_WIDTH,
      height: DEFAULT_BUTTON_HEIGHT,
      secondary: true,
      className: 'export-video-button',
      onClick: handlePreviewVideo
    }, "Preview"), React.createElement(ButtonGroup, null, React.createElement(Button, {
      width: DEFAULT_BUTTON_WIDTH,
      height: DEFAULT_BUTTON_HEIGHT,
      link: true,
      className: 'export-video-button',
      onClick: handleClose
    }, "Cancel"), React.createElement(Button, {
      width: DEFAULT_BUTTON_WIDTH,
      height: DEFAULT_BUTTON_HEIGHT,
      className: 'export-video-button',
      onClick: handleRenderVideo
    }, "Render")));
  });
};

export default withTheme(ExportVideoPanelFooter);
