// @hubble.gl/react/dist/esm/components/export-video/export-video-panel-settings.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/esm/taggedTemplateLiteral";

function _templateObject5() {
  var data = _taggedTemplateLiteral(["\n  display: grid;\n  grid-template-columns: 88px auto;\n  grid-template-rows: repeat(\n    ", ",\n    ", "\n  );\n  grid-row-gap: ", "px;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteral(["\n  align-self: center;\n  color: ", ";\n  font-weight: 500;\n  font-size: 11px;\n  padding: 0 12px;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n  align-self: center;\n  color: ", ";\n  font-weight: 400;\n  font-size: 11px;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n  position: relative;\n  flex-grow: 1;\n  margin-right: 24px;\n  margin-left: 24px;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  align-self: center;\n  color: ", ";\n  font-weight: 500;\n  font-size: 13px;\n  margin-top: ", ";\n  margin-bottom: ", "px;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import React from 'react';
import styled, { withTheme } from 'styled-components';
import { DEFAULT_PADDING, DEFAULT_ROW_GAP, DEFAULT_FILENAME, FORMATS, RESOLUTIONS } from './constants';
import { msConversion, estimateFileSize } from './utils';
import { WithKeplerUI } from '../inject-kepler';
var StyledSection = styled.div(_templateObject(), function (props) {
  return props.theme.labelColor;
}, function (props) {
  return props.top ? '0px' : "".concat(DEFAULT_PADDING, "px");
}, DEFAULT_ROW_GAP);
var SliderWrapper = styled.div(_templateObject2());
var StyledLabelCell = styled.div(_templateObject3(), function (props) {
  return props.theme.labelColor;
});
var StyledValueCell = styled.div(_templateObject4(), function (props) {
  return props.theme.textColor;
});
var InputGrid = styled.div(_templateObject5(), function (props) {
  return props.rows;
}, function (props) {
  return props.rowHeight ? props.rowHeight : '34px';
}, DEFAULT_ROW_GAP);

var getOptionValue = function getOptionValue(r) {
  return r.value;
};

var displayOption = function displayOption(r) {
  return r.label;
};

var getSelectedItems = function getSelectedItems(options, value) {
  return options.find(function (o) {
    return o.value === value;
  });
};

var ExportVideoPanelSettings = function ExportVideoPanelSettings(_ref) {
  var setMediaType = _ref.setMediaType,
      setCameraPreset = _ref.setCameraPreset,
      setFileName = _ref.setFileName,
      setResolution = _ref.setResolution,
      settingsData = _ref.settingsData,
      durationMs = _ref.durationMs,
      frameRate = _ref.frameRate,
      resolution = _ref.resolution,
      mediaType = _ref.mediaType,
      setDuration = _ref.setDuration;
  return React.createElement(WithKeplerUI, null, function (_ref2) {
    var Input = _ref2.Input,
        ItemSelector = _ref2.ItemSelector,
        Slider = _ref2.Slider;
    return React.createElement("div", null, React.createElement(StyledSection, {
      top: true
    }, "Export Settings"), React.createElement(InputGrid, {
      rows: 5
    }, React.createElement(StyledLabelCell, null, "File Name"), React.createElement(Input, {
      value: settingsData.fileName,
      placeholder: DEFAULT_FILENAME,
      onChange: function onChange(e) {
        return setFileName(e.target.value);
      }
    }), React.createElement(StyledLabelCell, null, "Media Type"), React.createElement(ItemSelector, {
      selectedItems: getSelectedItems(FORMATS, settingsData.mediaType),
      options: FORMATS,
      getOptionValue: getOptionValue,
      displayOption: displayOption,
      multiSelect: false,
      searchable: false,
      onChange: setMediaType
    }), React.createElement(StyledLabelCell, null, "Resolution"), React.createElement(ItemSelector, {
      selectedItems: getSelectedItems(RESOLUTIONS, settingsData.resolution),
      options: RESOLUTIONS,
      getOptionValue: getOptionValue,
      displayOption: displayOption,
      multiSelect: false,
      searchable: false,
      onChange: setResolution
    }), React.createElement(StyledLabelCell, null, "Duration"), React.createElement(StyledValueCell, null, React.createElement(SliderWrapper, {
      style: {
        width: '100%',
        marginLeft: '0px'
      },
      className: "modal-duration__slider"
    }, React.createElement(Slider, {
      showValues: false,
      enableBarDrag: true,
      isRanged: false,
      value1: durationMs,
      step: 100,
      minValue: 100,
      maxValue: 10000,
      style: {
        width: '70px'
      },
      onSlider1Change: function onSlider1Change(val) {
        setDuration(val);
      }
    }), React.createElement("div", {
      style: {
        alignSelf: 'center',
        paddingLeft: '8px',
        width: '56px'
      }
    }, msConversion(durationMs)))), React.createElement(StyledLabelCell, null, "File Size"), React.createElement(StyledValueCell, null, estimateFileSize(frameRate, resolution, durationMs, mediaType))), React.createElement(StyledSection, null, "Video Effects"), React.createElement(InputGrid, {
      rows: 1
    }, React.createElement(StyledLabelCell, null, "Camera"), React.createElement(ItemSelector, {
      selectedItems: settingsData.cameraPreset,
      options: ['None', 'Orbit (90ยบ)', 'Orbit (180ยบ)', 'Orbit (360ยบ)', 'North to South', 'South to North', 'East to West', 'West to East', 'Zoom Out', 'Zoom In'],
      multiSelect: false,
      searchable: false,
      onChange: setCameraPreset
    })));
  });
};

export default withTheme(ExportVideoPanelSettings);
