// @hubble.gl/react/dist/esm/components/export-video/export-video-panel.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/esm/taggedTemplateLiteral";

function _templateObject4() {
  var data = _taggedTemplateLiteral(["\n  width: ", "px;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n  padding: 0 ", "px;\n  display: grid;\n  grid-template-columns: ", "px ", "px;\n  grid-template-rows: auto;\n  grid-column-gap: ", "px;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  color: ", ";\n  font-size: 20px;\n  font-weight: 400;\n  line-height: ", ";\n  padding: 0 ", "px 16px ", "px;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n  justify-content: flex-end;\n  padding: ", "px ", "px 0 ", "px;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import React from 'react';
import styled, { withTheme } from 'styled-components';
import { DEFAULT_PADDING, DEFAULT_ICON_BUTTON_HEIGHT, DEFAULT_ROW_GAP, DEFAULT_SETTINGS_WIDTH } from './constants';
import ExportVideoPanelSettings from './export-video-panel-settings';
import { ExportVideoPanelPreview } from './export-video-panel-preview';
import ExportVideoPanelFooter from './export-video-panel-footer';
import { WithKeplerUI } from '../inject-kepler';
var PanelCloseInner = styled.div(_templateObject(), DEFAULT_PADDING, DEFAULT_PADDING, DEFAULT_PADDING);

var PanelClose = function PanelClose(_ref) {
  var handleClose = _ref.handleClose;
  return React.createElement(WithKeplerUI, null, function (_ref2) {
    var IconButton = _ref2.IconButton,
        Icons = _ref2.Icons;
    return React.createElement(PanelCloseInner, {
      className: "export-video-panel__close"
    }, React.createElement(IconButton, {
      className: "export-video-panel__button",
      link: true,
      onClick: handleClose
    }, React.createElement(Icons.Delete, {
      height: DEFAULT_ICON_BUTTON_HEIGHT
    })));
  });
};

var StyledTitle = styled.div(_templateObject2(), function (props) {
  return props.theme.titleTextColor;
}, function (props) {
  return props.theme.lineHeight;
}, DEFAULT_PADDING, DEFAULT_PADDING);
var PanelBodyInner = styled.div(_templateObject3(), DEFAULT_PADDING, function (props) {
  return props.exportVideoWidth;
}, DEFAULT_SETTINGS_WIDTH, DEFAULT_ROW_GAP);

var PanelBody = function PanelBody(_ref3) {
  var exportVideoWidth = _ref3.exportVideoWidth,
      mapData = _ref3.mapData,
      adapter = _ref3.adapter,
      setViewState = _ref3.setViewState,
      setMediaType = _ref3.setMediaType,
      setCameraPreset = _ref3.setCameraPreset,
      setFileName = _ref3.setFileName,
      setResolution = _ref3.setResolution,
      settingsData = _ref3.settingsData,
      durationMs = _ref3.durationMs,
      frameRate = _ref3.frameRate,
      resolution = _ref3.resolution,
      mediaType = _ref3.mediaType,
      viewState = _ref3.viewState,
      setDuration = _ref3.setDuration;
  return React.createElement(PanelBodyInner, {
    className: "export-video-panel__body",
    exportVideoWidth: exportVideoWidth
  }, React.createElement(ExportVideoPanelPreview, {
    mapData: mapData,
    adapter: adapter,
    setViewState: setViewState,
    exportVideoWidth: exportVideoWidth,
    resolution: resolution,
    viewState: viewState
  }), React.createElement(ExportVideoPanelSettings, {
    setMediaType: setMediaType,
    setCameraPreset: setCameraPreset,
    setFileName: setFileName,
    setResolution: setResolution,
    settingsData: settingsData,
    durationMs: durationMs,
    frameRate: frameRate,
    resolution: resolution,
    mediaType: mediaType,
    setDuration: setDuration
  }));
};

var Panel = styled.div(_templateObject4(), function (props) {
  return props.exportVideoWidth + 2 * DEFAULT_PADDING + DEFAULT_ROW_GAP + DEFAULT_SETTINGS_WIDTH;
});

var ExportVideoPanel = function ExportVideoPanel(_ref4) {
  var exportVideoWidth = _ref4.exportVideoWidth,
      handleClose = _ref4.handleClose,
      header = _ref4.header,
      mapData = _ref4.mapData,
      setViewState = _ref4.setViewState,
      settingsData = _ref4.settingsData,
      setMediaType = _ref4.setMediaType,
      setCameraPreset = _ref4.setCameraPreset,
      setFileName = _ref4.setFileName,
      setResolution = _ref4.setResolution,
      adapter = _ref4.adapter,
      handlePreviewVideo = _ref4.handlePreviewVideo,
      handleRenderVideo = _ref4.handleRenderVideo,
      durationMs = _ref4.durationMs,
      frameRate = _ref4.frameRate,
      resolution = _ref4.resolution,
      mediaType = _ref4.mediaType,
      viewState = _ref4.viewState,
      setDuration = _ref4.setDuration;
  return React.createElement(Panel, {
    exportVideoWidth: exportVideoWidth,
    className: "export-video-panel"
  }, header !== false ? React.createElement(React.Fragment, null, React.createElement(PanelClose, {
    handleClose: handleClose
  }), React.createElement(StyledTitle, {
    className: "export-video-panel__title"
  }, "Export Video")) : null, React.createElement(PanelBody, {
    exportVideoWidth: exportVideoWidth,
    mapData: mapData,
    adapter: adapter,
    setMediaType: setMediaType,
    setCameraPreset: setCameraPreset,
    setFileName: setFileName,
    setResolution: setResolution,
    settingsData: settingsData,
    setViewState: setViewState,
    durationMs: durationMs,
    frameRate: frameRate,
    resolution: resolution,
    mediaType: mediaType,
    viewState: viewState,
    setDuration: setDuration
  }), React.createElement(ExportVideoPanelFooter, {
    handleClose: handleClose,
    handlePreviewVideo: handlePreviewVideo,
    handleRenderVideo: handleRenderVideo
  }));
};

export default withTheme(ExportVideoPanel);
