// @kepler.gl/components/dist/common/file-uploader/upload-button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

var Wrapper = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: inline-block;\n  color: ", ";\n  font-size: 12px;\n  text-decoration: underline;\n\n  :hover {\n    cursor: pointer;\n    font-weight: 500;\n  }\n"])), function (props) {
  return props.theme.textColorLT;
});

var inputStyle = {
  display: 'none'
};

/*
Inspired by https://github.com/okonet/react-dropzone/blob/master/src/index.js
*/

/** @type {typeof import('./upload-button').UploadButton} */
var UploadButton = function UploadButton(_ref) {
  var onUpload = _ref.onUpload,
      children = _ref.children;

  var _fileInput = (0, _react.useRef)(null);

  var _onClick = (0, _react.useCallback)(function () {
    if (_fileInput.current) {
      // @ts-ignore create ref with useRef<HTMLInputElement>
      _fileInput.current.value = null; // @ts-ignore create ref with useRef<HTMLInputElement>

      _fileInput.current.click();
    }
  }, [_fileInput]);

  var _onChange = (0, _react.useCallback)(function (event) {
    var files = event.target.files;

    if (!files) {
      return;
    }

    onUpload(files, event);
  }, [onUpload]);

  return /*#__PURE__*/_react["default"].createElement(Wrapper, {
    className: "upload-button"
  }, /*#__PURE__*/_react["default"].createElement("input", {
    type: "file",
    ref: _fileInput,
    style: inputStyle,
    onChange: _onChange,
    className: "upload-button-input"
  }), /*#__PURE__*/_react["default"].createElement("span", {
    className: "file-upload__upload-button-span",
    onClick: _onClick
  }, children));
};

var _default = UploadButton;
exports["default"] = _default;
